{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ebe3ee85-87b3-4157-b1a4-1b0ba9f26d0f",
   "metadata": {},
   "source": [
    "*This Notebook provides a way to get rid of the data batch-effect. As a result of the work, a gene usage tables are formed. The [source code](https://github.com/antigenomics/mirpy) was kindly provided by our supervisors [Mikhail Shugai](https://github.com/mikessh) and [Elizaveta Vlasova](https://github.com/VEK239).*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "62a82f15-813f-4f78-9715-86a5dd920418",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "import re\n",
    "import time\n",
    "\n",
    "sys.path.append(\"../\")\n",
    "\n",
    "from mir.common.parser import *\n",
    "from mir.common.repertoire import Repertoire\n",
    "from mir.common.repertoire_dataset import RepertoireDataset\n",
    "from mir.basic.segment_usage import *\n",
    "from mir.basic.sampling import RepertoireSampling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b30a0c4-e796-412b-9d4e-13f2e2832e8d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "90ba1803-43d7-417e-82f8-0d9061ec23a3",
   "metadata": {},
   "source": [
    "# Cohort-I\n",
    "*fmba, Russia*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b785c5c-b35a-43f6-b841-a78bf3c9f1cf",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Alpha chain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "148b4d8c-56dc-4049-89d4-b5dd806f432b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file.id_short</th>\n",
       "      <th>sample.age</th>\n",
       "      <th>sample.sex</th>\n",
       "      <th>sample.height(cm)</th>\n",
       "      <th>sample.weight(kg)</th>\n",
       "      <th>sample.waist(cm)</th>\n",
       "      <th>sample.diastolicBloodPressure(mmHg)</th>\n",
       "      <th>sample.bodyTemperature</th>\n",
       "      <th>sample.pulse(beats/min)</th>\n",
       "      <th>sample.oxygenLevel(%)</th>\n",
       "      <th>...</th>\n",
       "      <th>sample.HLA-DRB1.1</th>\n",
       "      <th>sample.HLA-DRB1.2</th>\n",
       "      <th>Unnamed: 48</th>\n",
       "      <th>Unnamed: 49</th>\n",
       "      <th>Unnamed: 50</th>\n",
       "      <th>Unnamed: 51</th>\n",
       "      <th>Unnamed: 52</th>\n",
       "      <th>Unnamed: 53</th>\n",
       "      <th>run</th>\n",
       "      <th>folder</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>20000200808</td>\n",
       "      <td>19</td>\n",
       "      <td>male</td>\n",
       "      <td>187.0</td>\n",
       "      <td>105.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>80</td>\n",
       "      <td>36.6</td>\n",
       "      <td>74</td>\n",
       "      <td>99.0</td>\n",
       "      <td>...</td>\n",
       "      <td>DRB1*01:01</td>\n",
       "      <td>DRB1*04:04</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>020000200808_S181_L002.clonotypes.TRA.txt</td>\n",
       "      <td>2020/10_FMBA_NovaSeq5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>20000350808</td>\n",
       "      <td>54</td>\n",
       "      <td>female</td>\n",
       "      <td>159.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>90</td>\n",
       "      <td>36.3</td>\n",
       "      <td>60</td>\n",
       "      <td>98.0</td>\n",
       "      <td>...</td>\n",
       "      <td>DRB1*07:01</td>\n",
       "      <td>DRB1*07:01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>020000350808_S126_L002.clonotypes.TRA.txt</td>\n",
       "      <td>2020/10_FMBA_NovaSeq5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>20000360808</td>\n",
       "      <td>30</td>\n",
       "      <td>male</td>\n",
       "      <td>186.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>89.0</td>\n",
       "      <td>80</td>\n",
       "      <td>36.4</td>\n",
       "      <td>80</td>\n",
       "      <td>97.0</td>\n",
       "      <td>...</td>\n",
       "      <td>DRB1*03:01</td>\n",
       "      <td>DRB1*07:01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>020000360808_S102_L002.clonotypes.TRA.txt</td>\n",
       "      <td>2020/10_FMBA_NovaSeq5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>20000390808</td>\n",
       "      <td>33</td>\n",
       "      <td>female</td>\n",
       "      <td>157.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>60</td>\n",
       "      <td>36.6</td>\n",
       "      <td>72</td>\n",
       "      <td>97.0</td>\n",
       "      <td>...</td>\n",
       "      <td>DRB1*03:01</td>\n",
       "      <td>DRB1*12:02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>020000390808_S189_L002.clonotypes.TRA.txt</td>\n",
       "      <td>2020/10_FMBA_NovaSeq5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>20000640808</td>\n",
       "      <td>68</td>\n",
       "      <td>female</td>\n",
       "      <td>161.0</td>\n",
       "      <td>103.0</td>\n",
       "      <td>115.0</td>\n",
       "      <td>80</td>\n",
       "      <td>36.6</td>\n",
       "      <td>75</td>\n",
       "      <td>99.0</td>\n",
       "      <td>...</td>\n",
       "      <td>DRB1*01:01</td>\n",
       "      <td>DRB1*15:01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>020000640808_S117_L002.clonotypes.TRA.txt</td>\n",
       "      <td>2020/10_FMBA_NovaSeq5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>780003510808</td>\n",
       "      <td>61</td>\n",
       "      <td>male</td>\n",
       "      <td>176.0</td>\n",
       "      <td>85.0</td>\n",
       "      <td>106.0</td>\n",
       "      <td>80</td>\n",
       "      <td>36.4</td>\n",
       "      <td>60</td>\n",
       "      <td>99.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>780003510808_S92_L001.clonotypes.TRA.txt</td>\n",
       "      <td>2020/10_FMBA_NovaSeq4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>780003520808</td>\n",
       "      <td>81</td>\n",
       "      <td>male</td>\n",
       "      <td>172.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>78</td>\n",
       "      <td>36.5</td>\n",
       "      <td>70</td>\n",
       "      <td>99.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>780003520808_S12_L001.clonotypes.TRA.txt</td>\n",
       "      <td>2020/10_FMBA_NovaSeq4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>780003890808</td>\n",
       "      <td>27</td>\n",
       "      <td>female</td>\n",
       "      <td>168.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>70</td>\n",
       "      <td>36.5</td>\n",
       "      <td>64</td>\n",
       "      <td>99.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>780003890808_S35_L001.clonotypes.TRA.txt</td>\n",
       "      <td>2020/10_FMBA_NovaSeq4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>780003930808</td>\n",
       "      <td>64</td>\n",
       "      <td>male</td>\n",
       "      <td>171.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>89.0</td>\n",
       "      <td>75</td>\n",
       "      <td>36.6</td>\n",
       "      <td>72</td>\n",
       "      <td>99.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>780003930808_S76_L001.clonotypes.TRA.txt</td>\n",
       "      <td>2020/10_FMBA_NovaSeq4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>780004090808</td>\n",
       "      <td>34</td>\n",
       "      <td>female</td>\n",
       "      <td>164.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>70</td>\n",
       "      <td>36.6</td>\n",
       "      <td>74</td>\n",
       "      <td>99.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>780004090808_S60_L001.clonotypes.TRA.txt</td>\n",
       "      <td>2020/10_FMBA_NovaSeq4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>237 rows × 56 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     file.id_short  sample.age sample.sex  sample.height(cm)  \\\n",
       "0      20000200808          19       male              187.0   \n",
       "1      20000350808          54     female              159.0   \n",
       "2      20000360808          30       male              186.0   \n",
       "3      20000390808          33     female              157.0   \n",
       "4      20000640808          68     female              161.0   \n",
       "..             ...         ...        ...                ...   \n",
       "232   780003510808          61       male              176.0   \n",
       "233   780003520808          81       male              172.0   \n",
       "234   780003890808          27     female              168.0   \n",
       "235   780003930808          64       male              171.0   \n",
       "236   780004090808          34     female              164.0   \n",
       "\n",
       "     sample.weight(kg)  sample.waist(cm)  sample.diastolicBloodPressure(mmHg)  \\\n",
       "0                105.0              97.0                                   80   \n",
       "1                 86.0              96.0                                   90   \n",
       "2                 86.0              89.0                                   80   \n",
       "3                 65.0              77.0                                   60   \n",
       "4                103.0             115.0                                   80   \n",
       "..                 ...               ...                                  ...   \n",
       "232               85.0             106.0                                   80   \n",
       "233               74.0              96.0                                   78   \n",
       "234               61.0              59.0                                   70   \n",
       "235               60.0              89.0                                   75   \n",
       "236               58.0              61.0                                   70   \n",
       "\n",
       "     sample.bodyTemperature  sample.pulse(beats/min)  sample.oxygenLevel(%)  \\\n",
       "0                      36.6                       74                   99.0   \n",
       "1                      36.3                       60                   98.0   \n",
       "2                      36.4                       80                   97.0   \n",
       "3                      36.6                       72                   97.0   \n",
       "4                      36.6                       75                   99.0   \n",
       "..                      ...                      ...                    ...   \n",
       "232                    36.4                       60                   99.0   \n",
       "233                    36.5                       70                   99.0   \n",
       "234                    36.5                       64                   99.0   \n",
       "235                    36.6                       72                   99.0   \n",
       "236                    36.6                       74                   99.0   \n",
       "\n",
       "     ...  sample.HLA-DRB1.1  sample.HLA-DRB1.2  Unnamed: 48  Unnamed: 49  \\\n",
       "0    ...         DRB1*01:01         DRB1*04:04          NaN          NaN   \n",
       "1    ...         DRB1*07:01         DRB1*07:01          NaN          NaN   \n",
       "2    ...         DRB1*03:01         DRB1*07:01          NaN          NaN   \n",
       "3    ...         DRB1*03:01         DRB1*12:02          NaN          NaN   \n",
       "4    ...         DRB1*01:01         DRB1*15:01          NaN          NaN   \n",
       "..   ...                ...                ...          ...          ...   \n",
       "232  ...                NaN                NaN          NaN          NaN   \n",
       "233  ...                NaN                NaN          NaN          NaN   \n",
       "234  ...                NaN                NaN          NaN          NaN   \n",
       "235  ...                NaN                NaN          NaN          NaN   \n",
       "236  ...                NaN                NaN          NaN          NaN   \n",
       "\n",
       "     Unnamed: 50  Unnamed: 51  Unnamed: 52  Unnamed: 53  \\\n",
       "0            NaN          NaN          NaN          NaN   \n",
       "1            NaN          NaN          NaN          NaN   \n",
       "2            NaN          NaN          NaN          NaN   \n",
       "3            NaN          NaN          NaN          NaN   \n",
       "4            NaN          NaN          NaN          NaN   \n",
       "..           ...          ...          ...          ...   \n",
       "232          NaN          NaN          NaN          NaN   \n",
       "233          NaN          NaN          NaN          NaN   \n",
       "234          NaN          NaN          NaN          NaN   \n",
       "235          NaN          NaN          NaN          NaN   \n",
       "236          NaN          NaN          NaN          NaN   \n",
       "\n",
       "                                           run                 folder  \n",
       "0    020000200808_S181_L002.clonotypes.TRA.txt  2020/10_FMBA_NovaSeq5  \n",
       "1    020000350808_S126_L002.clonotypes.TRA.txt  2020/10_FMBA_NovaSeq5  \n",
       "2    020000360808_S102_L002.clonotypes.TRA.txt  2020/10_FMBA_NovaSeq5  \n",
       "3    020000390808_S189_L002.clonotypes.TRA.txt  2020/10_FMBA_NovaSeq5  \n",
       "4    020000640808_S117_L002.clonotypes.TRA.txt  2020/10_FMBA_NovaSeq5  \n",
       "..                                         ...                    ...  \n",
       "232   780003510808_S92_L001.clonotypes.TRA.txt  2020/10_FMBA_NovaSeq4  \n",
       "233   780003520808_S12_L001.clonotypes.TRA.txt  2020/10_FMBA_NovaSeq4  \n",
       "234   780003890808_S35_L001.clonotypes.TRA.txt  2020/10_FMBA_NovaSeq4  \n",
       "235   780003930808_S76_L001.clonotypes.TRA.txt  2020/10_FMBA_NovaSeq4  \n",
       "236   780004090808_S60_L001.clonotypes.TRA.txt  2020/10_FMBA_NovaSeq4  \n",
       "\n",
       "[237 rows x 56 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#metadata\n",
    "metadata_path = '../../../data/cohort_I/metadata/metadata_fmba_full.txt'\n",
    "batch_meta_path = '../../../data/cohort_I/metadata/desc_fmba_not_nan_hla.csv'\n",
    "\n",
    "metadata = pd.read_csv(metadata_path, sep='\\t')\n",
    "metadata.rename(columns={'id': 'file.id_short'}, inplace=True)\n",
    "metadata['file.id_short'] = metadata['file.id_short'].astype(int)\n",
    "\n",
    "batch_meta = pd.read_csv(batch_meta_path, sep=',')\n",
    "batch_meta['file.id_short'] = batch_meta['file.id'].str.split('_').str[0]\n",
    "batch_meta['file.id_short'] = batch_meta['file.id_short'].astype(int)\n",
    "batch_meta['run'] = batch_meta['file.id'] + '.clonotypes.TRA.txt'\n",
    "\n",
    "new_meta = pd.merge(metadata, batch_meta[['file.id_short', 'run', 'folder']], on='file.id_short')\n",
    "new_meta = new_meta[new_meta['sample.COVID_status'] == 'healthy']\n",
    "new_meta = new_meta.reset_index(drop=True)\n",
    "new_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e0c675fe-962c-4299-a6c0-cc9cc2769cc3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3574192\n",
      "500.78504848480225\n"
     ]
    }
   ],
   "source": [
    "base_path = '../../../data/cohort_I/data/'\n",
    "t0 = time.time()\n",
    "dataset = RepertoireDataset.load(parser=VDJtoolsParser(sep=','), \n",
    "                                 metadata=new_meta,\n",
    "                                 threads=28,\n",
    "                                 gene = 'TRA',\n",
    "                                 paths=[base_path + r['run']\n",
    "                                        for _, r in new_meta.iterrows()])\n",
    "print(time.time() - t0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7aa1cc1c-da70-4ade-8471-c961e197b9e4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file.id_short</th>\n",
       "      <th>sample.age</th>\n",
       "      <th>sample.sex</th>\n",
       "      <th>sample.height(cm)</th>\n",
       "      <th>sample.weight(kg)</th>\n",
       "      <th>sample.waist(cm)</th>\n",
       "      <th>sample.diastolicBloodPressure(mmHg)</th>\n",
       "      <th>sample.bodyTemperature</th>\n",
       "      <th>sample.pulse(beats/min)</th>\n",
       "      <th>sample.oxygenLevel(%)</th>\n",
       "      <th>...</th>\n",
       "      <th>sample.HLA-DRB1.2</th>\n",
       "      <th>Unnamed: 48</th>\n",
       "      <th>Unnamed: 49</th>\n",
       "      <th>Unnamed: 50</th>\n",
       "      <th>Unnamed: 51</th>\n",
       "      <th>Unnamed: 52</th>\n",
       "      <th>Unnamed: 53</th>\n",
       "      <th>run</th>\n",
       "      <th>folder</th>\n",
       "      <th>path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>20000200808</td>\n",
       "      <td>19</td>\n",
       "      <td>male</td>\n",
       "      <td>187.0</td>\n",
       "      <td>105.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>80</td>\n",
       "      <td>36.6</td>\n",
       "      <td>74</td>\n",
       "      <td>99.0</td>\n",
       "      <td>...</td>\n",
       "      <td>DRB1*04:04</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>020000200808_S181_L002.clonotypes.TRA.txt</td>\n",
       "      <td>2020/10_FMBA_NovaSeq5</td>\n",
       "      <td>../../../data/cohort_I/data/020000200808_S181_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>20000350808</td>\n",
       "      <td>54</td>\n",
       "      <td>female</td>\n",
       "      <td>159.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>90</td>\n",
       "      <td>36.3</td>\n",
       "      <td>60</td>\n",
       "      <td>98.0</td>\n",
       "      <td>...</td>\n",
       "      <td>DRB1*07:01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>020000350808_S126_L002.clonotypes.TRA.txt</td>\n",
       "      <td>2020/10_FMBA_NovaSeq5</td>\n",
       "      <td>../../../data/cohort_I/data/020000350808_S126_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>20000360808</td>\n",
       "      <td>30</td>\n",
       "      <td>male</td>\n",
       "      <td>186.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>89.0</td>\n",
       "      <td>80</td>\n",
       "      <td>36.4</td>\n",
       "      <td>80</td>\n",
       "      <td>97.0</td>\n",
       "      <td>...</td>\n",
       "      <td>DRB1*07:01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>020000360808_S102_L002.clonotypes.TRA.txt</td>\n",
       "      <td>2020/10_FMBA_NovaSeq5</td>\n",
       "      <td>../../../data/cohort_I/data/020000360808_S102_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>20000390808</td>\n",
       "      <td>33</td>\n",
       "      <td>female</td>\n",
       "      <td>157.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>60</td>\n",
       "      <td>36.6</td>\n",
       "      <td>72</td>\n",
       "      <td>97.0</td>\n",
       "      <td>...</td>\n",
       "      <td>DRB1*12:02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>020000390808_S189_L002.clonotypes.TRA.txt</td>\n",
       "      <td>2020/10_FMBA_NovaSeq5</td>\n",
       "      <td>../../../data/cohort_I/data/020000390808_S189_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>20000640808</td>\n",
       "      <td>68</td>\n",
       "      <td>female</td>\n",
       "      <td>161.0</td>\n",
       "      <td>103.0</td>\n",
       "      <td>115.0</td>\n",
       "      <td>80</td>\n",
       "      <td>36.6</td>\n",
       "      <td>75</td>\n",
       "      <td>99.0</td>\n",
       "      <td>...</td>\n",
       "      <td>DRB1*15:01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>020000640808_S117_L002.clonotypes.TRA.txt</td>\n",
       "      <td>2020/10_FMBA_NovaSeq5</td>\n",
       "      <td>../../../data/cohort_I/data/020000640808_S117_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>780003510808</td>\n",
       "      <td>61</td>\n",
       "      <td>male</td>\n",
       "      <td>176.0</td>\n",
       "      <td>85.0</td>\n",
       "      <td>106.0</td>\n",
       "      <td>80</td>\n",
       "      <td>36.4</td>\n",
       "      <td>60</td>\n",
       "      <td>99.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>780003510808_S92_L001.clonotypes.TRA.txt</td>\n",
       "      <td>2020/10_FMBA_NovaSeq4</td>\n",
       "      <td>../../../data/cohort_I/data/780003510808_S92_L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>780003520808</td>\n",
       "      <td>81</td>\n",
       "      <td>male</td>\n",
       "      <td>172.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>78</td>\n",
       "      <td>36.5</td>\n",
       "      <td>70</td>\n",
       "      <td>99.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>780003520808_S12_L001.clonotypes.TRA.txt</td>\n",
       "      <td>2020/10_FMBA_NovaSeq4</td>\n",
       "      <td>../../../data/cohort_I/data/780003520808_S12_L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>780003890808</td>\n",
       "      <td>27</td>\n",
       "      <td>female</td>\n",
       "      <td>168.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>70</td>\n",
       "      <td>36.5</td>\n",
       "      <td>64</td>\n",
       "      <td>99.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>780003890808_S35_L001.clonotypes.TRA.txt</td>\n",
       "      <td>2020/10_FMBA_NovaSeq4</td>\n",
       "      <td>../../../data/cohort_I/data/780003890808_S35_L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>780003930808</td>\n",
       "      <td>64</td>\n",
       "      <td>male</td>\n",
       "      <td>171.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>89.0</td>\n",
       "      <td>75</td>\n",
       "      <td>36.6</td>\n",
       "      <td>72</td>\n",
       "      <td>99.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>780003930808_S76_L001.clonotypes.TRA.txt</td>\n",
       "      <td>2020/10_FMBA_NovaSeq4</td>\n",
       "      <td>../../../data/cohort_I/data/780003930808_S76_L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>780004090808</td>\n",
       "      <td>34</td>\n",
       "      <td>female</td>\n",
       "      <td>164.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>70</td>\n",
       "      <td>36.6</td>\n",
       "      <td>74</td>\n",
       "      <td>99.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>780004090808_S60_L001.clonotypes.TRA.txt</td>\n",
       "      <td>2020/10_FMBA_NovaSeq4</td>\n",
       "      <td>../../../data/cohort_I/data/780004090808_S60_L...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>237 rows × 57 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     file.id_short  sample.age sample.sex  sample.height(cm)  \\\n",
       "0      20000200808          19       male              187.0   \n",
       "1      20000350808          54     female              159.0   \n",
       "2      20000360808          30       male              186.0   \n",
       "3      20000390808          33     female              157.0   \n",
       "4      20000640808          68     female              161.0   \n",
       "..             ...         ...        ...                ...   \n",
       "232   780003510808          61       male              176.0   \n",
       "233   780003520808          81       male              172.0   \n",
       "234   780003890808          27     female              168.0   \n",
       "235   780003930808          64       male              171.0   \n",
       "236   780004090808          34     female              164.0   \n",
       "\n",
       "     sample.weight(kg)  sample.waist(cm)  sample.diastolicBloodPressure(mmHg)  \\\n",
       "0                105.0              97.0                                   80   \n",
       "1                 86.0              96.0                                   90   \n",
       "2                 86.0              89.0                                   80   \n",
       "3                 65.0              77.0                                   60   \n",
       "4                103.0             115.0                                   80   \n",
       "..                 ...               ...                                  ...   \n",
       "232               85.0             106.0                                   80   \n",
       "233               74.0              96.0                                   78   \n",
       "234               61.0              59.0                                   70   \n",
       "235               60.0              89.0                                   75   \n",
       "236               58.0              61.0                                   70   \n",
       "\n",
       "     sample.bodyTemperature  sample.pulse(beats/min)  sample.oxygenLevel(%)  \\\n",
       "0                      36.6                       74                   99.0   \n",
       "1                      36.3                       60                   98.0   \n",
       "2                      36.4                       80                   97.0   \n",
       "3                      36.6                       72                   97.0   \n",
       "4                      36.6                       75                   99.0   \n",
       "..                      ...                      ...                    ...   \n",
       "232                    36.4                       60                   99.0   \n",
       "233                    36.5                       70                   99.0   \n",
       "234                    36.5                       64                   99.0   \n",
       "235                    36.6                       72                   99.0   \n",
       "236                    36.6                       74                   99.0   \n",
       "\n",
       "     ...  sample.HLA-DRB1.2  Unnamed: 48  Unnamed: 49  Unnamed: 50  \\\n",
       "0    ...         DRB1*04:04          NaN          NaN          NaN   \n",
       "1    ...         DRB1*07:01          NaN          NaN          NaN   \n",
       "2    ...         DRB1*07:01          NaN          NaN          NaN   \n",
       "3    ...         DRB1*12:02          NaN          NaN          NaN   \n",
       "4    ...         DRB1*15:01          NaN          NaN          NaN   \n",
       "..   ...                ...          ...          ...          ...   \n",
       "232  ...                NaN          NaN          NaN          NaN   \n",
       "233  ...                NaN          NaN          NaN          NaN   \n",
       "234  ...                NaN          NaN          NaN          NaN   \n",
       "235  ...                NaN          NaN          NaN          NaN   \n",
       "236  ...                NaN          NaN          NaN          NaN   \n",
       "\n",
       "     Unnamed: 51  Unnamed: 52  Unnamed: 53  \\\n",
       "0            NaN          NaN          NaN   \n",
       "1            NaN          NaN          NaN   \n",
       "2            NaN          NaN          NaN   \n",
       "3            NaN          NaN          NaN   \n",
       "4            NaN          NaN          NaN   \n",
       "..           ...          ...          ...   \n",
       "232          NaN          NaN          NaN   \n",
       "233          NaN          NaN          NaN   \n",
       "234          NaN          NaN          NaN   \n",
       "235          NaN          NaN          NaN   \n",
       "236          NaN          NaN          NaN   \n",
       "\n",
       "                                           run                 folder  \\\n",
       "0    020000200808_S181_L002.clonotypes.TRA.txt  2020/10_FMBA_NovaSeq5   \n",
       "1    020000350808_S126_L002.clonotypes.TRA.txt  2020/10_FMBA_NovaSeq5   \n",
       "2    020000360808_S102_L002.clonotypes.TRA.txt  2020/10_FMBA_NovaSeq5   \n",
       "3    020000390808_S189_L002.clonotypes.TRA.txt  2020/10_FMBA_NovaSeq5   \n",
       "4    020000640808_S117_L002.clonotypes.TRA.txt  2020/10_FMBA_NovaSeq5   \n",
       "..                                         ...                    ...   \n",
       "232   780003510808_S92_L001.clonotypes.TRA.txt  2020/10_FMBA_NovaSeq4   \n",
       "233   780003520808_S12_L001.clonotypes.TRA.txt  2020/10_FMBA_NovaSeq4   \n",
       "234   780003890808_S35_L001.clonotypes.TRA.txt  2020/10_FMBA_NovaSeq4   \n",
       "235   780003930808_S76_L001.clonotypes.TRA.txt  2020/10_FMBA_NovaSeq4   \n",
       "236   780004090808_S60_L001.clonotypes.TRA.txt  2020/10_FMBA_NovaSeq4   \n",
       "\n",
       "                                                  path  \n",
       "0    ../../../data/cohort_I/data/020000200808_S181_...  \n",
       "1    ../../../data/cohort_I/data/020000350808_S126_...  \n",
       "2    ../../../data/cohort_I/data/020000360808_S102_...  \n",
       "3    ../../../data/cohort_I/data/020000390808_S189_...  \n",
       "4    ../../../data/cohort_I/data/020000640808_S117_...  \n",
       "..                                                 ...  \n",
       "232  ../../../data/cohort_I/data/780003510808_S92_L...  \n",
       "233  ../../../data/cohort_I/data/780003520808_S12_L...  \n",
       "234  ../../../data/cohort_I/data/780003890808_S35_L...  \n",
       "235  ../../../data/cohort_I/data/780003930808_S76_L...  \n",
       "236  ../../../data/cohort_I/data/780004090808_S60_L...  \n",
       "\n",
       "[237 rows x 57 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset.metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c6e423f9-c4ae-4464-8281-908e8753a6e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TRAV35*01</th>\n",
       "      <th>TRAJ43*01</th>\n",
       "      <th>TRAJ8*01</th>\n",
       "      <th>TRAJ26*01</th>\n",
       "      <th>TRAJ25*01</th>\n",
       "      <th>TRAV8-6*01</th>\n",
       "      <th>TRAJ27*01</th>\n",
       "      <th>TRAJ19*01</th>\n",
       "      <th>TRAV9-2*01</th>\n",
       "      <th>TRAV1-1*01</th>\n",
       "      <th>...</th>\n",
       "      <th>TRAV17*01</th>\n",
       "      <th>TRAJ6*01</th>\n",
       "      <th>TRAJ21*01</th>\n",
       "      <th>TRAV19*01</th>\n",
       "      <th>TRAV18*01</th>\n",
       "      <th>TRAV12-1*01</th>\n",
       "      <th>TRAV9-1*01</th>\n",
       "      <th>TRAV26-1*01</th>\n",
       "      <th>TRAJ9*01</th>\n",
       "      <th>TRAJ61*01</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1112</td>\n",
       "      <td>1277</td>\n",
       "      <td>197</td>\n",
       "      <td>27</td>\n",
       "      <td>19</td>\n",
       "      <td>1006</td>\n",
       "      <td>557</td>\n",
       "      <td>0</td>\n",
       "      <td>1664</td>\n",
       "      <td>330</td>\n",
       "      <td>...</td>\n",
       "      <td>1463</td>\n",
       "      <td>624</td>\n",
       "      <td>3</td>\n",
       "      <td>1409</td>\n",
       "      <td>1</td>\n",
       "      <td>1336</td>\n",
       "      <td>17</td>\n",
       "      <td>1559</td>\n",
       "      <td>554</td>\n",
       "      <td>537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>791</td>\n",
       "      <td>500</td>\n",
       "      <td>282</td>\n",
       "      <td>27</td>\n",
       "      <td>13</td>\n",
       "      <td>833</td>\n",
       "      <td>493</td>\n",
       "      <td>0</td>\n",
       "      <td>1613</td>\n",
       "      <td>342</td>\n",
       "      <td>...</td>\n",
       "      <td>1292</td>\n",
       "      <td>628</td>\n",
       "      <td>2</td>\n",
       "      <td>996</td>\n",
       "      <td>1</td>\n",
       "      <td>1186</td>\n",
       "      <td>7</td>\n",
       "      <td>1250</td>\n",
       "      <td>608</td>\n",
       "      <td>276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1152</td>\n",
       "      <td>887</td>\n",
       "      <td>226</td>\n",
       "      <td>25</td>\n",
       "      <td>9</td>\n",
       "      <td>1017</td>\n",
       "      <td>578</td>\n",
       "      <td>0</td>\n",
       "      <td>1682</td>\n",
       "      <td>383</td>\n",
       "      <td>...</td>\n",
       "      <td>1532</td>\n",
       "      <td>821</td>\n",
       "      <td>2</td>\n",
       "      <td>1266</td>\n",
       "      <td>1</td>\n",
       "      <td>1481</td>\n",
       "      <td>15</td>\n",
       "      <td>1271</td>\n",
       "      <td>790</td>\n",
       "      <td>475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>857</td>\n",
       "      <td>706</td>\n",
       "      <td>203</td>\n",
       "      <td>25</td>\n",
       "      <td>15</td>\n",
       "      <td>701</td>\n",
       "      <td>419</td>\n",
       "      <td>1</td>\n",
       "      <td>1164</td>\n",
       "      <td>356</td>\n",
       "      <td>...</td>\n",
       "      <td>1257</td>\n",
       "      <td>479</td>\n",
       "      <td>1</td>\n",
       "      <td>843</td>\n",
       "      <td>0</td>\n",
       "      <td>950</td>\n",
       "      <td>9</td>\n",
       "      <td>874</td>\n",
       "      <td>447</td>\n",
       "      <td>453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>72</td>\n",
       "      <td>64</td>\n",
       "      <td>15</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>67</td>\n",
       "      <td>40</td>\n",
       "      <td>0</td>\n",
       "      <td>115</td>\n",
       "      <td>34</td>\n",
       "      <td>...</td>\n",
       "      <td>135</td>\n",
       "      <td>36</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>97</td>\n",
       "      <td>1</td>\n",
       "      <td>107</td>\n",
       "      <td>56</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>97</td>\n",
       "      <td>67</td>\n",
       "      <td>19</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>150</td>\n",
       "      <td>93</td>\n",
       "      <td>0</td>\n",
       "      <td>256</td>\n",
       "      <td>81</td>\n",
       "      <td>...</td>\n",
       "      <td>242</td>\n",
       "      <td>118</td>\n",
       "      <td>0</td>\n",
       "      <td>137</td>\n",
       "      <td>0</td>\n",
       "      <td>216</td>\n",
       "      <td>2</td>\n",
       "      <td>161</td>\n",
       "      <td>108</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>489</td>\n",
       "      <td>692</td>\n",
       "      <td>113</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>613</td>\n",
       "      <td>261</td>\n",
       "      <td>0</td>\n",
       "      <td>990</td>\n",
       "      <td>191</td>\n",
       "      <td>...</td>\n",
       "      <td>941</td>\n",
       "      <td>432</td>\n",
       "      <td>1</td>\n",
       "      <td>521</td>\n",
       "      <td>0</td>\n",
       "      <td>662</td>\n",
       "      <td>0</td>\n",
       "      <td>930</td>\n",
       "      <td>368</td>\n",
       "      <td>165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>131</td>\n",
       "      <td>118</td>\n",
       "      <td>30</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>132</td>\n",
       "      <td>61</td>\n",
       "      <td>0</td>\n",
       "      <td>247</td>\n",
       "      <td>38</td>\n",
       "      <td>...</td>\n",
       "      <td>230</td>\n",
       "      <td>80</td>\n",
       "      <td>0</td>\n",
       "      <td>164</td>\n",
       "      <td>0</td>\n",
       "      <td>177</td>\n",
       "      <td>1</td>\n",
       "      <td>151</td>\n",
       "      <td>80</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>261</td>\n",
       "      <td>175</td>\n",
       "      <td>29</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>236</td>\n",
       "      <td>142</td>\n",
       "      <td>0</td>\n",
       "      <td>445</td>\n",
       "      <td>86</td>\n",
       "      <td>...</td>\n",
       "      <td>371</td>\n",
       "      <td>186</td>\n",
       "      <td>0</td>\n",
       "      <td>238</td>\n",
       "      <td>0</td>\n",
       "      <td>325</td>\n",
       "      <td>1</td>\n",
       "      <td>439</td>\n",
       "      <td>170</td>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>269</td>\n",
       "      <td>224</td>\n",
       "      <td>67</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>240</td>\n",
       "      <td>194</td>\n",
       "      <td>0</td>\n",
       "      <td>573</td>\n",
       "      <td>113</td>\n",
       "      <td>...</td>\n",
       "      <td>552</td>\n",
       "      <td>196</td>\n",
       "      <td>0</td>\n",
       "      <td>430</td>\n",
       "      <td>0</td>\n",
       "      <td>370</td>\n",
       "      <td>1</td>\n",
       "      <td>471</td>\n",
       "      <td>188</td>\n",
       "      <td>131</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>237 rows × 109 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     TRAV35*01  TRAJ43*01  TRAJ8*01  TRAJ26*01  TRAJ25*01  TRAV8-6*01  \\\n",
       "0         1112       1277       197         27         19        1006   \n",
       "1          791        500       282         27         13         833   \n",
       "2         1152        887       226         25          9        1017   \n",
       "3          857        706       203         25         15         701   \n",
       "4           72         64        15          4          2          67   \n",
       "..         ...        ...       ...        ...        ...         ...   \n",
       "232         97         67        19          4          2         150   \n",
       "233        489        692       113          5          5         613   \n",
       "234        131        118        30          6          2         132   \n",
       "235        261        175        29          1          0         236   \n",
       "236        269        224        67          6          2         240   \n",
       "\n",
       "     TRAJ27*01  TRAJ19*01  TRAV9-2*01  TRAV1-1*01  ...  TRAV17*01  TRAJ6*01  \\\n",
       "0          557          0        1664         330  ...       1463       624   \n",
       "1          493          0        1613         342  ...       1292       628   \n",
       "2          578          0        1682         383  ...       1532       821   \n",
       "3          419          1        1164         356  ...       1257       479   \n",
       "4           40          0         115          34  ...        135        36   \n",
       "..         ...        ...         ...         ...  ...        ...       ...   \n",
       "232         93          0         256          81  ...        242       118   \n",
       "233        261          0         990         191  ...        941       432   \n",
       "234         61          0         247          38  ...        230        80   \n",
       "235        142          0         445          86  ...        371       186   \n",
       "236        194          0         573         113  ...        552       196   \n",
       "\n",
       "     TRAJ21*01  TRAV19*01  TRAV18*01  TRAV12-1*01  TRAV9-1*01  TRAV26-1*01  \\\n",
       "0            3       1409          1         1336          17         1559   \n",
       "1            2        996          1         1186           7         1250   \n",
       "2            2       1266          1         1481          15         1271   \n",
       "3            1        843          0          950           9          874   \n",
       "4            0         60          0           97           1          107   \n",
       "..         ...        ...        ...          ...         ...          ...   \n",
       "232          0        137          0          216           2          161   \n",
       "233          1        521          0          662           0          930   \n",
       "234          0        164          0          177           1          151   \n",
       "235          0        238          0          325           1          439   \n",
       "236          0        430          0          370           1          471   \n",
       "\n",
       "     TRAJ9*01  TRAJ61*01  \n",
       "0         554        537  \n",
       "1         608        276  \n",
       "2         790        475  \n",
       "3         447        453  \n",
       "4          56         12  \n",
       "..        ...        ...  \n",
       "232       108         41  \n",
       "233       368        165  \n",
       "234        80         64  \n",
       "235       170         69  \n",
       "236       188        131  \n",
       "\n",
       "[237 rows x 109 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset.evaluate_segment_usage()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c991103f-82fb-47ff-986a-47d8ef1822a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "folder_to_run_mapping = {}\n",
    "for folder in dataset.metadata[['run', 'folder']].folder.unique():\n",
    "    folder_to_run_mapping[folder] = set(dataset.metadata[dataset.metadata.folder == folder].run)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6ddd1e44-1da0-470e-8f05-381b48fbf922",
   "metadata": {},
   "outputs": [],
   "source": [
    "norm_usage_table_v = NormalizedSegmentUsageTable.load_from_repertoire_dataset(repertoire_dataset=dataset, gene='TRA', segment_type='V')\n",
    "norm_usage_table_j = NormalizedSegmentUsageTable.load_from_repertoire_dataset(repertoire_dataset=dataset, gene='TRA', segment_type='J')\n",
    "\n",
    "z_score_usage_table_v = StandardizedSegmentUsageTable.load_from_repertoire_dataset(repertoire_dataset=dataset, gene='TRA', segment_type='V', group_mapping=folder_to_run_mapping, standardization_method='log_exp')\n",
    "z_score_usage_table_j = StandardizedSegmentUsageTable.load_from_repertoire_dataset(repertoire_dataset=dataset, gene='TRA', segment_type='J', group_mapping=folder_to_run_mapping, standardization_method='log_exp')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3f4d010a-bd5a-4f6a-96fa-0a970f55ae59",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "     TRAV35*01  TRAV8-6*01  TRAV9-2*01  TRAV1-1*01  TRAV22*01  TRAV36DV7*01  \\\n",
       "0     0.029492    0.026681    0.044132    0.008752   0.020130      0.015515   \n",
       "1     0.024952    0.026277    0.050882    0.010788   0.016813      0.012713   \n",
       "2     0.029591    0.026123    0.043205    0.009838   0.019496      0.014513   \n",
       "3     0.029603    0.024214    0.040207    0.012297   0.020276      0.013264   \n",
       "4     0.027896    0.025959    0.044556    0.013173   0.019760      0.014336   \n",
       "..         ...         ...         ...         ...        ...           ...   \n",
       "232   0.018979    0.029348    0.050088    0.015848   0.016239      0.009783   \n",
       "233   0.025247    0.031649    0.051113    0.009861   0.017451      0.012597   \n",
       "234   0.026039    0.026237    0.049096    0.007553   0.016895      0.014311   \n",
       "235   0.030877    0.027919    0.052644    0.010174   0.024015      0.003076   \n",
       "236   0.023387    0.020866    0.049817    0.009824   0.018779      0.017649   \n",
       "\n",
       "     TRAV13-1*01  TRAV6*01  TRAV8-4*01  TRAV21*01  ...  TRAV12-2*01  \\\n",
       "0       0.041347  0.015727    0.022650   0.034346  ...     0.030314   \n",
       "1       0.046497  0.017381    0.032333   0.043942  ...     0.023122   \n",
       "2       0.044617  0.016285    0.028178   0.039737  ...     0.032879   \n",
       "3       0.043040  0.017306    0.029016   0.038964  ...     0.031883   \n",
       "4       0.058892  0.022472    0.031383   0.031383  ...     0.036420   \n",
       "..           ...       ...         ...        ...  ...          ...   \n",
       "232     0.046762  0.019761    0.036588   0.040110  ...     0.027783   \n",
       "233     0.051009  0.015024    0.035159   0.029015  ...     0.032939   \n",
       "234     0.052872  0.019479    0.028623   0.032995  ...     0.027430   \n",
       "235     0.047794  0.018455    0.018692   0.035964  ...     0.031823   \n",
       "236     0.049122  0.017823    0.030256   0.039732  ...     0.028256   \n",
       "\n",
       "     TRAV25*01  TRAV26-2*01  TRAV2*01  TRAV17*01  TRAV19*01  TRAV18*01  \\\n",
       "0     0.022782     0.028246  0.015144   0.038801   0.037369   0.000027   \n",
       "1     0.019305     0.021450  0.023469   0.040756   0.031419   0.000032   \n",
       "2     0.019624     0.026996  0.023118   0.039352   0.032519   0.000026   \n",
       "3     0.019551     0.022798  0.017168   0.043420   0.029119   0.000000   \n",
       "4     0.021697     0.020922  0.022084   0.052305   0.023247   0.000000   \n",
       "..         ...          ...       ...        ...        ...        ...   \n",
       "232   0.021718     0.023087  0.028175   0.047349   0.026805   0.000000   \n",
       "233   0.018999     0.024782  0.016263   0.048583   0.026899   0.000000   \n",
       "234   0.016895     0.023057  0.022262   0.045717   0.032598   0.000000   \n",
       "235   0.022714     0.029220  0.015616   0.043890   0.028156   0.000000   \n",
       "236   0.019736     0.030343  0.022431   0.047992   0.037385   0.000000   \n",
       "\n",
       "     TRAV12-1*01  TRAV9-1*01  TRAV26-1*01  \n",
       "0       0.035433    0.000451     0.041347  \n",
       "1       0.037412    0.000221     0.039431  \n",
       "2       0.038042    0.000385     0.032648  \n",
       "3       0.032815    0.000311     0.030190  \n",
       "4       0.037582    0.000387     0.041457  \n",
       "..           ...         ...          ...  \n",
       "232     0.042262    0.000391     0.031501  \n",
       "233     0.034178    0.000000     0.048015  \n",
       "234     0.035182    0.000199     0.030014  \n",
       "235     0.038448    0.000118     0.051934  \n",
       "236     0.032168    0.000087     0.040949  \n",
       "\n",
       "[237 rows x 48 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "norm_usage_table_v"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ea75a9d-d752-4ad4-a2e4-a503094c5c79",
   "metadata": {},
   "source": [
    "#### Sampling reads from the repertoire, not changing the gene usage distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "aec0049c-5928-42f3-935e-30c43c3ee305",
   "metadata": {},
   "outputs": [],
   "source": [
    "sampled_repertoire = RepertoireSampling().sample(repertoire=dataset[0], old_usage_matrix=[norm_usage_table_v, norm_usage_table_j], new_usage_matrix=None, n=100000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "088f3544-0cff-4606-9c3b-6f2f2cec94d8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Repertoire of 21868 clonotypes and 100000 cells:\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CAVRDG*_QGGKLIF cdr3nt_0\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CAATDSWGKLQF cdr3nt_1\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CVVSPLW_NNRLAF cdr3nt_2\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CASVDAGNMLTF cdr3nt_3\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CATGPTR_GTALIF cdr3nt_4\n",
       "file.id_short                                                                20000200808\n",
       "sample.age                                                                            19\n",
       "sample.sex                                                                          male\n",
       "sample.height(cm)                                                                  187.0\n",
       "sample.weight(kg)                                                                  105.0\n",
       "sample.waist(cm)                                                                    97.0\n",
       "sample.diastolicBloodPressure(mmHg)                                                   80\n",
       "sample.bodyTemperature                                                              36.6\n",
       "sample.pulse(beats/min)                                                               74\n",
       "sample.oxygenLevel(%)                                                               99.0\n",
       "sample.erythrocytes(10*12/l)                                                        5.03\n",
       "sample.hemoglobinCount(g/l)                                                        172.0\n",
       "sample.hematocrit(%)                                                                50.0\n",
       "sample.plateletCount(10*9/l)                                                       256.0\n",
       "sample.reticulocyteCount(10*9/l)                                                     NaN\n",
       "sample.WBC(10*9/l)                                                                  10.7\n",
       "sample.ESR(mm/h)                                                                     4.0\n",
       "sample.serumTotalProtein(g/l)                                                       78.8\n",
       "sample.hemoglobin(g/l)                                                               NaN\n",
       "sample.COVID_PCR                                                                negative\n",
       "sample.bloodType                                                                  B(III)\n",
       "sample.RhFactor                                                                 poditive\n",
       "sample.COVID_IgM                                                                      no\n",
       "sample.COVID_IgG                                                                      no\n",
       "sample.COVID_status                                                              healthy\n",
       "sample.neutrophil(10*9/l)                                                            7.2\n",
       "sample.lymphocytes(10*9/l)                                                           2.7\n",
       "sample.monocyte(10*9/l)                                                              0.8\n",
       "sample.eosinophils(10*9/l)                                                           NaN\n",
       "sample.basophil(10*9/l)                                                              NaN\n",
       "sample.NLR                                                                           NaN\n",
       "sample.neutrophilFraction                                                           69.0\n",
       "sample.lymphocytesFraction                                                          24.0\n",
       "sample.monocyteFraction                                                              6.0\n",
       "sample.eosinophilsFraction                                                           1.0\n",
       "sample.basophilFraction                                                              0.0\n",
       "sample.HLA-A.1                                                                   A*02:01\n",
       "sample.HLA-A.2                                                                   A*11:01\n",
       "sample.HLA-B.1                                                                   B*40:01\n",
       "sample.HLA-B.2                                                                   B*52:01\n",
       "sample.HLA-C.1                                                                   C*03:04\n",
       "sample.HLA-C.2                                                                   C*12:02\n",
       "sample.HLA-DPB1.1                                                             DPB1*04:01\n",
       "sample.HLA-DPB1.2                                                             DPB1*06:01\n",
       "sample.HLA-DQB1.1                                                             DQB1*03:02\n",
       "sample.HLA-DQB1.2                                                             DQB1*05:01\n",
       "sample.HLA-DRB1.1                                                             DRB1*01:01\n",
       "sample.HLA-DRB1.2                                                             DRB1*04:04\n",
       "Unnamed: 48                                                                          NaN\n",
       "Unnamed: 49                                                                          NaN\n",
       "Unnamed: 50                                                                          NaN\n",
       "Unnamed: 51                                                                          NaN\n",
       "Unnamed: 52                                                                          NaN\n",
       "Unnamed: 53                                                                          NaN\n",
       "run                                            020000200808_S181_L002.clonotypes.TRA.txt\n",
       "folder                                                             2020/10_FMBA_NovaSeq5\n",
       "path                                   ../../../data/cohort_I/data/020000200808_S181_...\n",
       "Name: 0, dtype: object\n",
       "..."
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampled_repertoire"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e05d983-3a64-42ef-b7bc-88c19c7354fd",
   "metadata": {},
   "source": [
    "#### Resampling the repertoire using the new derived usages for V and J genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7f0deef6-8df8-40b5-9679-acb7ad9a8e06",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6.547260761260986\n"
     ]
    }
   ],
   "source": [
    "t0 = time.time()\n",
    "resampled_repertoire = RepertoireSampling().sample(repertoire=dataset[0], old_usage_matrix=[norm_usage_table_v, norm_usage_table_j], new_usage_matrix=[z_score_usage_table_v, z_score_usage_table_j])\n",
    "print(time.time() - t0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f07c3120-349d-4e1c-bc22-04268c6cf65b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Repertoire of 36002 clonotypes and 9537751 cells:\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CAVRDG*_QGGKLIF cdr3nt_0\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CAATDSWGKLQF cdr3nt_1\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CATGPTR_GTALIF cdr3nt_4\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CVVSPLW_NNRLAF cdr3nt_2\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CAVSDGWGNTPLVF cdr3nt_10\n",
       "file.id_short                                                                20000200808\n",
       "sample.age                                                                            19\n",
       "sample.sex                                                                          male\n",
       "sample.height(cm)                                                                  187.0\n",
       "sample.weight(kg)                                                                  105.0\n",
       "sample.waist(cm)                                                                    97.0\n",
       "sample.diastolicBloodPressure(mmHg)                                                   80\n",
       "sample.bodyTemperature                                                              36.6\n",
       "sample.pulse(beats/min)                                                               74\n",
       "sample.oxygenLevel(%)                                                               99.0\n",
       "sample.erythrocytes(10*12/l)                                                        5.03\n",
       "sample.hemoglobinCount(g/l)                                                        172.0\n",
       "sample.hematocrit(%)                                                                50.0\n",
       "sample.plateletCount(10*9/l)                                                       256.0\n",
       "sample.reticulocyteCount(10*9/l)                                                     NaN\n",
       "sample.WBC(10*9/l)                                                                  10.7\n",
       "sample.ESR(mm/h)                                                                     4.0\n",
       "sample.serumTotalProtein(g/l)                                                       78.8\n",
       "sample.hemoglobin(g/l)                                                               NaN\n",
       "sample.COVID_PCR                                                                negative\n",
       "sample.bloodType                                                                  B(III)\n",
       "sample.RhFactor                                                                 poditive\n",
       "sample.COVID_IgM                                                                      no\n",
       "sample.COVID_IgG                                                                      no\n",
       "sample.COVID_status                                                              healthy\n",
       "sample.neutrophil(10*9/l)                                                            7.2\n",
       "sample.lymphocytes(10*9/l)                                                           2.7\n",
       "sample.monocyte(10*9/l)                                                              0.8\n",
       "sample.eosinophils(10*9/l)                                                           NaN\n",
       "sample.basophil(10*9/l)                                                              NaN\n",
       "sample.NLR                                                                           NaN\n",
       "sample.neutrophilFraction                                                           69.0\n",
       "sample.lymphocytesFraction                                                          24.0\n",
       "sample.monocyteFraction                                                              6.0\n",
       "sample.eosinophilsFraction                                                           1.0\n",
       "sample.basophilFraction                                                              0.0\n",
       "sample.HLA-A.1                                                                   A*02:01\n",
       "sample.HLA-A.2                                                                   A*11:01\n",
       "sample.HLA-B.1                                                                   B*40:01\n",
       "sample.HLA-B.2                                                                   B*52:01\n",
       "sample.HLA-C.1                                                                   C*03:04\n",
       "sample.HLA-C.2                                                                   C*12:02\n",
       "sample.HLA-DPB1.1                                                             DPB1*04:01\n",
       "sample.HLA-DPB1.2                                                             DPB1*06:01\n",
       "sample.HLA-DQB1.1                                                             DQB1*03:02\n",
       "sample.HLA-DQB1.2                                                             DQB1*05:01\n",
       "sample.HLA-DRB1.1                                                             DRB1*01:01\n",
       "sample.HLA-DRB1.2                                                             DRB1*04:04\n",
       "Unnamed: 48                                                                          NaN\n",
       "Unnamed: 49                                                                          NaN\n",
       "Unnamed: 50                                                                          NaN\n",
       "Unnamed: 51                                                                          NaN\n",
       "Unnamed: 52                                                                          NaN\n",
       "Unnamed: 53                                                                          NaN\n",
       "run                                            020000200808_S181_L002.clonotypes.TRA.txt\n",
       "folder                                                             2020/10_FMBA_NovaSeq5\n",
       "path                                   ../../../data/cohort_I/data/020000200808_S181_...\n",
       "Name: 0, dtype: object\n",
       "..."
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resampled_repertoire"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d9217d7-f159-4886-911f-4457bc5b31fd",
   "metadata": {},
   "source": [
    "#### Resampling all the reperoires from a dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "de50d8da-a14d-4e4e-8f84-8f5269547ca1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3271399\n",
      "563.537807226181\n"
     ]
    }
   ],
   "source": [
    "t0 = time.time()\n",
    "resampled_dataset = dataset.resample(updated_segment_usage_tables=[z_score_usage_table_v, z_score_usage_table_j], threads=28)\n",
    "print(time.time() - t0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "dd3edc75-e17b-451e-acfa-90cde049eeb8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Repertoire of 29777 clonotypes and 6356503 cells:\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CAGQRFGGSQGNLIF cdr3nt_0\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CAGQTFYNQGGKLIF cdr3nt_1\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CLLGDPGGTSYGKLTF cdr3nt_2\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CAVSLQGGKLIF cdr3nt_3\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CAETPIYNQGGKLIF cdr3nt_4\n",
       "file.id_short                                                                50001340808\n",
       "sample.age                                                                            28\n",
       "sample.sex                                                                        female\n",
       "sample.height(cm)                                                                  160.0\n",
       "sample.weight(kg)                                                                   64.0\n",
       "sample.waist(cm)                                                                    67.0\n",
       "sample.diastolicBloodPressure(mmHg)                                                   60\n",
       "sample.bodyTemperature                                                              36.3\n",
       "sample.pulse(beats/min)                                                               64\n",
       "sample.oxygenLevel(%)                                                               98.0\n",
       "sample.erythrocytes(10*12/l)                                                        4.78\n",
       "sample.hemoglobinCount(g/l)                                                        116.0\n",
       "sample.hematocrit(%)                                                                36.9\n",
       "sample.plateletCount(10*9/l)                                                       386.0\n",
       "sample.reticulocyteCount(10*9/l)                                                   0.357\n",
       "sample.WBC(10*9/l)                                                                  4.24\n",
       "sample.ESR(mm/h)                                                                    11.0\n",
       "sample.serumTotalProtein(g/l)                                                       75.0\n",
       "sample.hemoglobin(g/l)                                                               NaN\n",
       "sample.COVID_PCR                                                                negative\n",
       "sample.bloodType                                                                   A(II)\n",
       "sample.RhFactor                                                                 poditive\n",
       "sample.COVID_IgM                                                                      no\n",
       "sample.COVID_IgG                                                                      no\n",
       "sample.COVID_status                                                              healthy\n",
       "sample.neutrophil(10*9/l)                                                            NaN\n",
       "sample.lymphocytes(10*9/l)                                                           NaN\n",
       "sample.monocyte(10*9/l)                                                              NaN\n",
       "sample.eosinophils(10*9/l)                                                           NaN\n",
       "sample.basophil(10*9/l)                                                              NaN\n",
       "sample.NLR                                                                           NaN\n",
       "sample.neutrophilFraction                                                            NaN\n",
       "sample.lymphocytesFraction                                                          31.5\n",
       "sample.monocyteFraction                                                              1.0\n",
       "sample.eosinophilsFraction                                                           NaN\n",
       "sample.basophilFraction                                                              NaN\n",
       "sample.HLA-A.1                                                                   A*02:01\n",
       "sample.HLA-A.2                                                                   A*32:01\n",
       "sample.HLA-B.1                                                                   B*07:02\n",
       "sample.HLA-B.2                                                                   B*37:01\n",
       "sample.HLA-C.1                                                                   C*02:02\n",
       "sample.HLA-C.2                                                                   C*07:02\n",
       "sample.HLA-DPB1.1                                                             DPB1*03:01\n",
       "sample.HLA-DPB1.2                                                             DPB1*04:01\n",
       "sample.HLA-DQB1.1                                                             DQB1*03:01\n",
       "sample.HLA-DQB1.2                                                             DQB1*06:02\n",
       "sample.HLA-DRB1.1                                                             DRB1*08:03\n",
       "sample.HLA-DRB1.2                                                             DRB1*15:01\n",
       "Unnamed: 48                                                                          NaN\n",
       "Unnamed: 49                                                                          NaN\n",
       "Unnamed: 50                                                                          NaN\n",
       "Unnamed: 51                                                                          NaN\n",
       "Unnamed: 52                                                                          NaN\n",
       "Unnamed: 53                                                                          NaN\n",
       "run                                             050001340808_S85_L001.clonotypes.TRA.txt\n",
       "folder                                                         2020/12_FMBA_Novaseq7_DNA\n",
       "path                                   ../../../data/cohort_I/data/050001340808_S85_L...\n",
       "Name: 14, dtype: object\n",
       "..."
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset[14]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ea7416df-0121-4d9d-8f01-ac67eb4a0095",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Repertoire of 28822 clonotypes and 6356503 cells:\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CLLGDPGGTSYGKLTF cdr3nt_2\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CAVFWGFNKFYF cdr3nt_349\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CAVNAPSNARLMF cdr3nt_5\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CVVSAS_DKFYF cdr3nt_410\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CVVSPYNFNKFYF cdr3nt_700\n",
       "file.id_short                                                                50001340808\n",
       "sample.age                                                                            28\n",
       "sample.sex                                                                        female\n",
       "sample.height(cm)                                                                  160.0\n",
       "sample.weight(kg)                                                                   64.0\n",
       "sample.waist(cm)                                                                    67.0\n",
       "sample.diastolicBloodPressure(mmHg)                                                   60\n",
       "sample.bodyTemperature                                                              36.3\n",
       "sample.pulse(beats/min)                                                               64\n",
       "sample.oxygenLevel(%)                                                               98.0\n",
       "sample.erythrocytes(10*12/l)                                                        4.78\n",
       "sample.hemoglobinCount(g/l)                                                        116.0\n",
       "sample.hematocrit(%)                                                                36.9\n",
       "sample.plateletCount(10*9/l)                                                       386.0\n",
       "sample.reticulocyteCount(10*9/l)                                                   0.357\n",
       "sample.WBC(10*9/l)                                                                  4.24\n",
       "sample.ESR(mm/h)                                                                    11.0\n",
       "sample.serumTotalProtein(g/l)                                                       75.0\n",
       "sample.hemoglobin(g/l)                                                               NaN\n",
       "sample.COVID_PCR                                                                negative\n",
       "sample.bloodType                                                                   A(II)\n",
       "sample.RhFactor                                                                 poditive\n",
       "sample.COVID_IgM                                                                      no\n",
       "sample.COVID_IgG                                                                      no\n",
       "sample.COVID_status                                                              healthy\n",
       "sample.neutrophil(10*9/l)                                                            NaN\n",
       "sample.lymphocytes(10*9/l)                                                           NaN\n",
       "sample.monocyte(10*9/l)                                                              NaN\n",
       "sample.eosinophils(10*9/l)                                                           NaN\n",
       "sample.basophil(10*9/l)                                                              NaN\n",
       "sample.NLR                                                                           NaN\n",
       "sample.neutrophilFraction                                                            NaN\n",
       "sample.lymphocytesFraction                                                          31.5\n",
       "sample.monocyteFraction                                                              1.0\n",
       "sample.eosinophilsFraction                                                           NaN\n",
       "sample.basophilFraction                                                              NaN\n",
       "sample.HLA-A.1                                                                   A*02:01\n",
       "sample.HLA-A.2                                                                   A*32:01\n",
       "sample.HLA-B.1                                                                   B*07:02\n",
       "sample.HLA-B.2                                                                   B*37:01\n",
       "sample.HLA-C.1                                                                   C*02:02\n",
       "sample.HLA-C.2                                                                   C*07:02\n",
       "sample.HLA-DPB1.1                                                             DPB1*03:01\n",
       "sample.HLA-DPB1.2                                                             DPB1*04:01\n",
       "sample.HLA-DQB1.1                                                             DQB1*03:01\n",
       "sample.HLA-DQB1.2                                                             DQB1*06:02\n",
       "sample.HLA-DRB1.1                                                             DRB1*08:03\n",
       "sample.HLA-DRB1.2                                                             DRB1*15:01\n",
       "Unnamed: 48                                                                          NaN\n",
       "Unnamed: 49                                                                          NaN\n",
       "Unnamed: 50                                                                          NaN\n",
       "Unnamed: 51                                                                          NaN\n",
       "Unnamed: 52                                                                          NaN\n",
       "Unnamed: 53                                                                          NaN\n",
       "run                                             050001340808_S85_L001.clonotypes.TRA.txt\n",
       "folder                                                         2020/12_FMBA_Novaseq7_DNA\n",
       "path                                   ../../../data/cohort_I/data/050001340808_S85_L...\n",
       "Name: 14, dtype: object\n",
       "..."
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resampled_dataset[14]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4e3ee864-d0b4-4999-ae66-542c431a85b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "trav_matrix = NormalizedSegmentUsageTable.load_from_repertoire_dataset(\n",
    "    resampled_dataset, gene='TRA', segment_type='V').segment_usage_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "9d2151c1-badc-4755-8e8b-49fbc53b2b80",
   "metadata": {},
   "outputs": [],
   "source": [
    "traj_matrix = NormalizedSegmentUsageTable.load_from_repertoire_dataset(\n",
    "    resampled_dataset, gene='TRA', segment_type='J').segment_usage_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "fb2a78df-23e7-4b9a-bc1c-2167201d8009",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TRAJ43*01</th>\n",
       "      <th>TRAJ8*01</th>\n",
       "      <th>TRAJ26*01</th>\n",
       "      <th>TRAJ25*01</th>\n",
       "      <th>TRAJ27*01</th>\n",
       "      <th>TRAJ19*01</th>\n",
       "      <th>TRAJ29*01</th>\n",
       "      <th>TRAJ28*01</th>\n",
       "      <th>TRAJ38*01</th>\n",
       "      <th>TRAJ18*01</th>\n",
       "      <th>...</th>\n",
       "      <th>TRAJ14*01</th>\n",
       "      <th>TRAJ40*01</th>\n",
       "      <th>TRAJ15*01</th>\n",
       "      <th>TRAJ41*01</th>\n",
       "      <th>TRAJ13*01</th>\n",
       "      <th>TRAJ57*01</th>\n",
       "      <th>TRAJ6*01</th>\n",
       "      <th>TRAJ21*01</th>\n",
       "      <th>TRAJ9*01</th>\n",
       "      <th>TRAJ61*01</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.033915</td>\n",
       "      <td>0.005055</td>\n",
       "      <td>0.000750</td>\n",
       "      <td>0.000389</td>\n",
       "      <td>0.015110</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.024582</td>\n",
       "      <td>0.024804</td>\n",
       "      <td>0.010638</td>\n",
       "      <td>0.009111</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000028</td>\n",
       "      <td>0.031332</td>\n",
       "      <td>0.018110</td>\n",
       "      <td>0.016971</td>\n",
       "      <td>0.017055</td>\n",
       "      <td>0.021693</td>\n",
       "      <td>0.016777</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.015249</td>\n",
       "      <td>0.014221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.016807</td>\n",
       "      <td>0.008690</td>\n",
       "      <td>0.000808</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>0.015999</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.028461</td>\n",
       "      <td>0.023206</td>\n",
       "      <td>0.009296</td>\n",
       "      <td>0.009700</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000168</td>\n",
       "      <td>0.027854</td>\n",
       "      <td>0.016908</td>\n",
       "      <td>0.013675</td>\n",
       "      <td>0.019434</td>\n",
       "      <td>0.022465</td>\n",
       "      <td>0.019906</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.019299</td>\n",
       "      <td>0.008454</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.023339</td>\n",
       "      <td>0.005713</td>\n",
       "      <td>0.000647</td>\n",
       "      <td>0.000189</td>\n",
       "      <td>0.015254</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.030831</td>\n",
       "      <td>0.023500</td>\n",
       "      <td>0.007681</td>\n",
       "      <td>0.009756</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>0.030534</td>\n",
       "      <td>0.017652</td>\n",
       "      <td>0.016817</td>\n",
       "      <td>0.018730</td>\n",
       "      <td>0.009405</td>\n",
       "      <td>0.021344</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.020428</td>\n",
       "      <td>0.012181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.025046</td>\n",
       "      <td>0.006808</td>\n",
       "      <td>0.000896</td>\n",
       "      <td>0.000466</td>\n",
       "      <td>0.014798</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.026766</td>\n",
       "      <td>0.026479</td>\n",
       "      <td>0.011036</td>\n",
       "      <td>0.008922</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.029811</td>\n",
       "      <td>0.015945</td>\n",
       "      <td>0.016733</td>\n",
       "      <td>0.018095</td>\n",
       "      <td>0.013042</td>\n",
       "      <td>0.016769</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.015766</td>\n",
       "      <td>0.015551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.025530</td>\n",
       "      <td>0.005193</td>\n",
       "      <td>0.001731</td>\n",
       "      <td>0.000865</td>\n",
       "      <td>0.015578</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.026828</td>\n",
       "      <td>0.023367</td>\n",
       "      <td>0.007356</td>\n",
       "      <td>0.007356</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.020770</td>\n",
       "      <td>0.023367</td>\n",
       "      <td>0.014712</td>\n",
       "      <td>0.025963</td>\n",
       "      <td>0.000865</td>\n",
       "      <td>0.015145</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.020770</td>\n",
       "      <td>0.005193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>0.012786</td>\n",
       "      <td>0.003773</td>\n",
       "      <td>0.000838</td>\n",
       "      <td>0.000419</td>\n",
       "      <td>0.019074</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.033117</td>\n",
       "      <td>0.024942</td>\n",
       "      <td>0.005869</td>\n",
       "      <td>0.011318</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.022427</td>\n",
       "      <td>0.019493</td>\n",
       "      <td>0.015720</td>\n",
       "      <td>0.021798</td>\n",
       "      <td>0.006078</td>\n",
       "      <td>0.024104</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.021798</td>\n",
       "      <td>0.008384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>0.036421</td>\n",
       "      <td>0.005876</td>\n",
       "      <td>0.000212</td>\n",
       "      <td>0.000159</td>\n",
       "      <td>0.013393</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.025040</td>\n",
       "      <td>0.021440</td>\n",
       "      <td>0.009317</td>\n",
       "      <td>0.009052</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000053</td>\n",
       "      <td>0.034569</td>\n",
       "      <td>0.017046</td>\n",
       "      <td>0.018317</td>\n",
       "      <td>0.020487</td>\n",
       "      <td>0.015352</td>\n",
       "      <td>0.022340</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.019111</td>\n",
       "      <td>0.008576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>0.023800</td>\n",
       "      <td>0.005647</td>\n",
       "      <td>0.001210</td>\n",
       "      <td>0.000403</td>\n",
       "      <td>0.012303</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.026624</td>\n",
       "      <td>0.021178</td>\n",
       "      <td>0.012505</td>\n",
       "      <td>0.010085</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000202</td>\n",
       "      <td>0.032271</td>\n",
       "      <td>0.019968</td>\n",
       "      <td>0.015329</td>\n",
       "      <td>0.017346</td>\n",
       "      <td>0.023598</td>\n",
       "      <td>0.016136</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.016136</td>\n",
       "      <td>0.012707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>0.021071</td>\n",
       "      <td>0.003512</td>\n",
       "      <td>0.000125</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.016681</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.025712</td>\n",
       "      <td>0.028722</td>\n",
       "      <td>0.003386</td>\n",
       "      <td>0.010410</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021824</td>\n",
       "      <td>0.021197</td>\n",
       "      <td>0.015552</td>\n",
       "      <td>0.019691</td>\n",
       "      <td>0.006522</td>\n",
       "      <td>0.022451</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.019691</td>\n",
       "      <td>0.008153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>0.019587</td>\n",
       "      <td>0.005055</td>\n",
       "      <td>0.000451</td>\n",
       "      <td>0.000090</td>\n",
       "      <td>0.017420</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.029515</td>\n",
       "      <td>0.022926</td>\n",
       "      <td>0.007853</td>\n",
       "      <td>0.008394</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.029245</td>\n",
       "      <td>0.014261</td>\n",
       "      <td>0.014351</td>\n",
       "      <td>0.014081</td>\n",
       "      <td>0.014171</td>\n",
       "      <td>0.016608</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.016337</td>\n",
       "      <td>0.011734</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>237 rows × 61 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     TRAJ43*01  TRAJ8*01  TRAJ26*01  TRAJ25*01  TRAJ27*01  TRAJ19*01  \\\n",
       "0     0.033915  0.005055   0.000750   0.000389   0.015110   0.000000   \n",
       "1     0.016807  0.008690   0.000808   0.000337   0.015999   0.000000   \n",
       "2     0.023339  0.005713   0.000647   0.000189   0.015254   0.000000   \n",
       "3     0.025046  0.006808   0.000896   0.000466   0.014798   0.000036   \n",
       "4     0.025530  0.005193   0.001731   0.000865   0.015578   0.000000   \n",
       "..         ...       ...        ...        ...        ...        ...   \n",
       "232   0.012786  0.003773   0.000838   0.000419   0.019074   0.000000   \n",
       "233   0.036421  0.005876   0.000212   0.000159   0.013393   0.000000   \n",
       "234   0.023800  0.005647   0.001210   0.000403   0.012303   0.000000   \n",
       "235   0.021071  0.003512   0.000125   0.000000   0.016681   0.000000   \n",
       "236   0.019587  0.005055   0.000451   0.000090   0.017420   0.000000   \n",
       "\n",
       "     TRAJ29*01  TRAJ28*01  TRAJ38*01  TRAJ18*01  ...  TRAJ14*01  TRAJ40*01  \\\n",
       "0     0.024582   0.024804   0.010638   0.009111  ...   0.000028   0.031332   \n",
       "1     0.028461   0.023206   0.009296   0.009700  ...   0.000168   0.027854   \n",
       "2     0.030831   0.023500   0.007681   0.009756  ...   0.000027   0.030534   \n",
       "3     0.026766   0.026479   0.011036   0.008922  ...   0.000036   0.029811   \n",
       "4     0.026828   0.023367   0.007356   0.007356  ...   0.000000   0.020770   \n",
       "..         ...        ...        ...        ...  ...        ...        ...   \n",
       "232   0.033117   0.024942   0.005869   0.011318  ...   0.000000   0.022427   \n",
       "233   0.025040   0.021440   0.009317   0.009052  ...   0.000053   0.034569   \n",
       "234   0.026624   0.021178   0.012505   0.010085  ...   0.000202   0.032271   \n",
       "235   0.025712   0.028722   0.003386   0.010410  ...   0.000000   0.021824   \n",
       "236   0.029515   0.022926   0.007853   0.008394  ...   0.000000   0.029245   \n",
       "\n",
       "     TRAJ15*01  TRAJ41*01  TRAJ13*01  TRAJ57*01  TRAJ6*01  TRAJ21*01  \\\n",
       "0     0.018110   0.016971   0.017055   0.021693  0.016777        0.0   \n",
       "1     0.016908   0.013675   0.019434   0.022465  0.019906        0.0   \n",
       "2     0.017652   0.016817   0.018730   0.009405  0.021344        0.0   \n",
       "3     0.015945   0.016733   0.018095   0.013042  0.016769        0.0   \n",
       "4     0.023367   0.014712   0.025963   0.000865  0.015145        0.0   \n",
       "..         ...        ...        ...        ...       ...        ...   \n",
       "232   0.019493   0.015720   0.021798   0.006078  0.024104        0.0   \n",
       "233   0.017046   0.018317   0.020487   0.015352  0.022340        0.0   \n",
       "234   0.019968   0.015329   0.017346   0.023598  0.016136        0.0   \n",
       "235   0.021197   0.015552   0.019691   0.006522  0.022451        0.0   \n",
       "236   0.014261   0.014351   0.014081   0.014171  0.016608        0.0   \n",
       "\n",
       "     TRAJ9*01  TRAJ61*01  \n",
       "0    0.015249   0.014221  \n",
       "1    0.019299   0.008454  \n",
       "2    0.020428   0.012181  \n",
       "3    0.015766   0.015551  \n",
       "4    0.020770   0.005193  \n",
       "..        ...        ...  \n",
       "232  0.021798   0.008384  \n",
       "233  0.019111   0.008576  \n",
       "234  0.016136   0.012707  \n",
       "235  0.019691   0.008153  \n",
       "236  0.016337   0.011734  \n",
       "\n",
       "[237 rows x 61 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "traj_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1d60f133-1388-473b-8937-120e2f7ec0be",
   "metadata": {},
   "outputs": [],
   "source": [
    "trav_matrix.to_csv('../../../data/cohort_I/data_corrected/v_clonotypes_TRA.csv', index = False)\n",
    "traj_matrix.to_csv('../../../data/cohort_I/data_corrected/j_clonotypes_TRA.csv', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14c01be8-0a0c-4c03-89df-1d3af63a1768",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "74a60589-6d1c-431a-845e-bcfa2b067a00",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Nonfunctional sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "19f7a449-a9e4-4e92-9b28-2080edccafac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "285.883665561676\n"
     ]
    }
   ],
   "source": [
    "base_path = '../../../data/cohort_I/nonfunctional/'\n",
    "t0 = time.time()\n",
    "dataset = RepertoireDataset.load(parser=VDJtoolsParser(sep=','), \n",
    "                                 metadata=new_meta,\n",
    "                                 threads=28,\n",
    "                                 gene = 'TRA',\n",
    "                                 paths=[base_path + r['run']\n",
    "                                        for _, r in new_meta.iterrows()])\n",
    "print(time.time() - t0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "776b20ee-e71e-4a3b-9a03-7ec25559b50e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file.id_short</th>\n",
       "      <th>sample.age</th>\n",
       "      <th>sample.sex</th>\n",
       "      <th>sample.height(cm)</th>\n",
       "      <th>sample.weight(kg)</th>\n",
       "      <th>sample.waist(cm)</th>\n",
       "      <th>sample.diastolicBloodPressure(mmHg)</th>\n",
       "      <th>sample.bodyTemperature</th>\n",
       "      <th>sample.pulse(beats/min)</th>\n",
       "      <th>sample.oxygenLevel(%)</th>\n",
       "      <th>...</th>\n",
       "      <th>sample.HLA-DRB1.2</th>\n",
       "      <th>Unnamed: 48</th>\n",
       "      <th>Unnamed: 49</th>\n",
       "      <th>Unnamed: 50</th>\n",
       "      <th>Unnamed: 51</th>\n",
       "      <th>Unnamed: 52</th>\n",
       "      <th>Unnamed: 53</th>\n",
       "      <th>run</th>\n",
       "      <th>folder</th>\n",
       "      <th>path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>20000200808</td>\n",
       "      <td>19</td>\n",
       "      <td>male</td>\n",
       "      <td>187.0</td>\n",
       "      <td>105.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>80</td>\n",
       "      <td>36.6</td>\n",
       "      <td>74</td>\n",
       "      <td>99.0</td>\n",
       "      <td>...</td>\n",
       "      <td>DRB1*04:04</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>020000200808_S181_L002.clonotypes.TRA.txt</td>\n",
       "      <td>2020/10_FMBA_NovaSeq5</td>\n",
       "      <td>../../../data/cohort_I/nonfunctional/020000200...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>20000350808</td>\n",
       "      <td>54</td>\n",
       "      <td>female</td>\n",
       "      <td>159.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>90</td>\n",
       "      <td>36.3</td>\n",
       "      <td>60</td>\n",
       "      <td>98.0</td>\n",
       "      <td>...</td>\n",
       "      <td>DRB1*07:01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>020000350808_S126_L002.clonotypes.TRA.txt</td>\n",
       "      <td>2020/10_FMBA_NovaSeq5</td>\n",
       "      <td>../../../data/cohort_I/nonfunctional/020000350...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>20000360808</td>\n",
       "      <td>30</td>\n",
       "      <td>male</td>\n",
       "      <td>186.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>89.0</td>\n",
       "      <td>80</td>\n",
       "      <td>36.4</td>\n",
       "      <td>80</td>\n",
       "      <td>97.0</td>\n",
       "      <td>...</td>\n",
       "      <td>DRB1*07:01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>020000360808_S102_L002.clonotypes.TRA.txt</td>\n",
       "      <td>2020/10_FMBA_NovaSeq5</td>\n",
       "      <td>../../../data/cohort_I/nonfunctional/020000360...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>20000390808</td>\n",
       "      <td>33</td>\n",
       "      <td>female</td>\n",
       "      <td>157.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>60</td>\n",
       "      <td>36.6</td>\n",
       "      <td>72</td>\n",
       "      <td>97.0</td>\n",
       "      <td>...</td>\n",
       "      <td>DRB1*12:02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>020000390808_S189_L002.clonotypes.TRA.txt</td>\n",
       "      <td>2020/10_FMBA_NovaSeq5</td>\n",
       "      <td>../../../data/cohort_I/nonfunctional/020000390...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>20000640808</td>\n",
       "      <td>68</td>\n",
       "      <td>female</td>\n",
       "      <td>161.0</td>\n",
       "      <td>103.0</td>\n",
       "      <td>115.0</td>\n",
       "      <td>80</td>\n",
       "      <td>36.6</td>\n",
       "      <td>75</td>\n",
       "      <td>99.0</td>\n",
       "      <td>...</td>\n",
       "      <td>DRB1*15:01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>020000640808_S117_L002.clonotypes.TRA.txt</td>\n",
       "      <td>2020/10_FMBA_NovaSeq5</td>\n",
       "      <td>../../../data/cohort_I/nonfunctional/020000640...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>780003510808</td>\n",
       "      <td>61</td>\n",
       "      <td>male</td>\n",
       "      <td>176.0</td>\n",
       "      <td>85.0</td>\n",
       "      <td>106.0</td>\n",
       "      <td>80</td>\n",
       "      <td>36.4</td>\n",
       "      <td>60</td>\n",
       "      <td>99.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>780003510808_S92_L001.clonotypes.TRA.txt</td>\n",
       "      <td>2020/10_FMBA_NovaSeq4</td>\n",
       "      <td>../../../data/cohort_I/nonfunctional/780003510...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>780003520808</td>\n",
       "      <td>81</td>\n",
       "      <td>male</td>\n",
       "      <td>172.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>78</td>\n",
       "      <td>36.5</td>\n",
       "      <td>70</td>\n",
       "      <td>99.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>780003520808_S12_L001.clonotypes.TRA.txt</td>\n",
       "      <td>2020/10_FMBA_NovaSeq4</td>\n",
       "      <td>../../../data/cohort_I/nonfunctional/780003520...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>780003890808</td>\n",
       "      <td>27</td>\n",
       "      <td>female</td>\n",
       "      <td>168.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>70</td>\n",
       "      <td>36.5</td>\n",
       "      <td>64</td>\n",
       "      <td>99.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>780003890808_S35_L001.clonotypes.TRA.txt</td>\n",
       "      <td>2020/10_FMBA_NovaSeq4</td>\n",
       "      <td>../../../data/cohort_I/nonfunctional/780003890...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>780003930808</td>\n",
       "      <td>64</td>\n",
       "      <td>male</td>\n",
       "      <td>171.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>89.0</td>\n",
       "      <td>75</td>\n",
       "      <td>36.6</td>\n",
       "      <td>72</td>\n",
       "      <td>99.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>780003930808_S76_L001.clonotypes.TRA.txt</td>\n",
       "      <td>2020/10_FMBA_NovaSeq4</td>\n",
       "      <td>../../../data/cohort_I/nonfunctional/780003930...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>780004090808</td>\n",
       "      <td>34</td>\n",
       "      <td>female</td>\n",
       "      <td>164.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>70</td>\n",
       "      <td>36.6</td>\n",
       "      <td>74</td>\n",
       "      <td>99.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>780004090808_S60_L001.clonotypes.TRA.txt</td>\n",
       "      <td>2020/10_FMBA_NovaSeq4</td>\n",
       "      <td>../../../data/cohort_I/nonfunctional/780004090...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>237 rows × 57 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     file.id_short  sample.age sample.sex  sample.height(cm)  \\\n",
       "0      20000200808          19       male              187.0   \n",
       "1      20000350808          54     female              159.0   \n",
       "2      20000360808          30       male              186.0   \n",
       "3      20000390808          33     female              157.0   \n",
       "4      20000640808          68     female              161.0   \n",
       "..             ...         ...        ...                ...   \n",
       "232   780003510808          61       male              176.0   \n",
       "233   780003520808          81       male              172.0   \n",
       "234   780003890808          27     female              168.0   \n",
       "235   780003930808          64       male              171.0   \n",
       "236   780004090808          34     female              164.0   \n",
       "\n",
       "     sample.weight(kg)  sample.waist(cm)  sample.diastolicBloodPressure(mmHg)  \\\n",
       "0                105.0              97.0                                   80   \n",
       "1                 86.0              96.0                                   90   \n",
       "2                 86.0              89.0                                   80   \n",
       "3                 65.0              77.0                                   60   \n",
       "4                103.0             115.0                                   80   \n",
       "..                 ...               ...                                  ...   \n",
       "232               85.0             106.0                                   80   \n",
       "233               74.0              96.0                                   78   \n",
       "234               61.0              59.0                                   70   \n",
       "235               60.0              89.0                                   75   \n",
       "236               58.0              61.0                                   70   \n",
       "\n",
       "     sample.bodyTemperature  sample.pulse(beats/min)  sample.oxygenLevel(%)  \\\n",
       "0                      36.6                       74                   99.0   \n",
       "1                      36.3                       60                   98.0   \n",
       "2                      36.4                       80                   97.0   \n",
       "3                      36.6                       72                   97.0   \n",
       "4                      36.6                       75                   99.0   \n",
       "..                      ...                      ...                    ...   \n",
       "232                    36.4                       60                   99.0   \n",
       "233                    36.5                       70                   99.0   \n",
       "234                    36.5                       64                   99.0   \n",
       "235                    36.6                       72                   99.0   \n",
       "236                    36.6                       74                   99.0   \n",
       "\n",
       "     ...  sample.HLA-DRB1.2  Unnamed: 48  Unnamed: 49  Unnamed: 50  \\\n",
       "0    ...         DRB1*04:04          NaN          NaN          NaN   \n",
       "1    ...         DRB1*07:01          NaN          NaN          NaN   \n",
       "2    ...         DRB1*07:01          NaN          NaN          NaN   \n",
       "3    ...         DRB1*12:02          NaN          NaN          NaN   \n",
       "4    ...         DRB1*15:01          NaN          NaN          NaN   \n",
       "..   ...                ...          ...          ...          ...   \n",
       "232  ...                NaN          NaN          NaN          NaN   \n",
       "233  ...                NaN          NaN          NaN          NaN   \n",
       "234  ...                NaN          NaN          NaN          NaN   \n",
       "235  ...                NaN          NaN          NaN          NaN   \n",
       "236  ...                NaN          NaN          NaN          NaN   \n",
       "\n",
       "     Unnamed: 51  Unnamed: 52  Unnamed: 53  \\\n",
       "0            NaN          NaN          NaN   \n",
       "1            NaN          NaN          NaN   \n",
       "2            NaN          NaN          NaN   \n",
       "3            NaN          NaN          NaN   \n",
       "4            NaN          NaN          NaN   \n",
       "..           ...          ...          ...   \n",
       "232          NaN          NaN          NaN   \n",
       "233          NaN          NaN          NaN   \n",
       "234          NaN          NaN          NaN   \n",
       "235          NaN          NaN          NaN   \n",
       "236          NaN          NaN          NaN   \n",
       "\n",
       "                                           run                 folder  \\\n",
       "0    020000200808_S181_L002.clonotypes.TRA.txt  2020/10_FMBA_NovaSeq5   \n",
       "1    020000350808_S126_L002.clonotypes.TRA.txt  2020/10_FMBA_NovaSeq5   \n",
       "2    020000360808_S102_L002.clonotypes.TRA.txt  2020/10_FMBA_NovaSeq5   \n",
       "3    020000390808_S189_L002.clonotypes.TRA.txt  2020/10_FMBA_NovaSeq5   \n",
       "4    020000640808_S117_L002.clonotypes.TRA.txt  2020/10_FMBA_NovaSeq5   \n",
       "..                                         ...                    ...   \n",
       "232   780003510808_S92_L001.clonotypes.TRA.txt  2020/10_FMBA_NovaSeq4   \n",
       "233   780003520808_S12_L001.clonotypes.TRA.txt  2020/10_FMBA_NovaSeq4   \n",
       "234   780003890808_S35_L001.clonotypes.TRA.txt  2020/10_FMBA_NovaSeq4   \n",
       "235   780003930808_S76_L001.clonotypes.TRA.txt  2020/10_FMBA_NovaSeq4   \n",
       "236   780004090808_S60_L001.clonotypes.TRA.txt  2020/10_FMBA_NovaSeq4   \n",
       "\n",
       "                                                  path  \n",
       "0    ../../../data/cohort_I/nonfunctional/020000200...  \n",
       "1    ../../../data/cohort_I/nonfunctional/020000350...  \n",
       "2    ../../../data/cohort_I/nonfunctional/020000360...  \n",
       "3    ../../../data/cohort_I/nonfunctional/020000390...  \n",
       "4    ../../../data/cohort_I/nonfunctional/020000640...  \n",
       "..                                                 ...  \n",
       "232  ../../../data/cohort_I/nonfunctional/780003510...  \n",
       "233  ../../../data/cohort_I/nonfunctional/780003520...  \n",
       "234  ../../../data/cohort_I/nonfunctional/780003890...  \n",
       "235  ../../../data/cohort_I/nonfunctional/780003930...  \n",
       "236  ../../../data/cohort_I/nonfunctional/780004090...  \n",
       "\n",
       "[237 rows x 57 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset.metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "9b57a4cd-04a9-482d-ba97-e7098d91c488",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TRAV35*01</th>\n",
       "      <th>TRAJ43*01</th>\n",
       "      <th>TRAJ8*01</th>\n",
       "      <th>TRAJ26*01</th>\n",
       "      <th>TRAJ25*01</th>\n",
       "      <th>TRAV8-6*01</th>\n",
       "      <th>TRAJ27*01</th>\n",
       "      <th>TRAJ19*01</th>\n",
       "      <th>TRAV9-2*01</th>\n",
       "      <th>TRAV1-1*01</th>\n",
       "      <th>...</th>\n",
       "      <th>TRAV17*01</th>\n",
       "      <th>TRAJ6*01</th>\n",
       "      <th>TRAJ21*01</th>\n",
       "      <th>TRAV19*01</th>\n",
       "      <th>TRAV18*01</th>\n",
       "      <th>TRAV12-1*01</th>\n",
       "      <th>TRAV9-1*01</th>\n",
       "      <th>TRAV26-1*01</th>\n",
       "      <th>TRAJ9*01</th>\n",
       "      <th>TRAJ61*01</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>634</td>\n",
       "      <td>615</td>\n",
       "      <td>61</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>377</td>\n",
       "      <td>269</td>\n",
       "      <td>0</td>\n",
       "      <td>663</td>\n",
       "      <td>178</td>\n",
       "      <td>...</td>\n",
       "      <td>785</td>\n",
       "      <td>278</td>\n",
       "      <td>1</td>\n",
       "      <td>634</td>\n",
       "      <td>0</td>\n",
       "      <td>646</td>\n",
       "      <td>8</td>\n",
       "      <td>632</td>\n",
       "      <td>187</td>\n",
       "      <td>369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>361</td>\n",
       "      <td>216</td>\n",
       "      <td>103</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>278</td>\n",
       "      <td>213</td>\n",
       "      <td>0</td>\n",
       "      <td>479</td>\n",
       "      <td>169</td>\n",
       "      <td>...</td>\n",
       "      <td>537</td>\n",
       "      <td>233</td>\n",
       "      <td>0</td>\n",
       "      <td>546</td>\n",
       "      <td>0</td>\n",
       "      <td>537</td>\n",
       "      <td>4</td>\n",
       "      <td>498</td>\n",
       "      <td>166</td>\n",
       "      <td>180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>605</td>\n",
       "      <td>470</td>\n",
       "      <td>82</td>\n",
       "      <td>8</td>\n",
       "      <td>4</td>\n",
       "      <td>382</td>\n",
       "      <td>279</td>\n",
       "      <td>0</td>\n",
       "      <td>671</td>\n",
       "      <td>227</td>\n",
       "      <td>...</td>\n",
       "      <td>835</td>\n",
       "      <td>350</td>\n",
       "      <td>1</td>\n",
       "      <td>646</td>\n",
       "      <td>1</td>\n",
       "      <td>685</td>\n",
       "      <td>10</td>\n",
       "      <td>633</td>\n",
       "      <td>268</td>\n",
       "      <td>306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>466</td>\n",
       "      <td>382</td>\n",
       "      <td>89</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>303</td>\n",
       "      <td>213</td>\n",
       "      <td>1</td>\n",
       "      <td>499</td>\n",
       "      <td>162</td>\n",
       "      <td>...</td>\n",
       "      <td>626</td>\n",
       "      <td>212</td>\n",
       "      <td>1</td>\n",
       "      <td>495</td>\n",
       "      <td>0</td>\n",
       "      <td>498</td>\n",
       "      <td>7</td>\n",
       "      <td>462</td>\n",
       "      <td>157</td>\n",
       "      <td>291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>32</td>\n",
       "      <td>24</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>35</td>\n",
       "      <td>21</td>\n",
       "      <td>...</td>\n",
       "      <td>55</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>43</td>\n",
       "      <td>1</td>\n",
       "      <td>44</td>\n",
       "      <td>15</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>52</td>\n",
       "      <td>28</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>34</td>\n",
       "      <td>40</td>\n",
       "      <td>0</td>\n",
       "      <td>91</td>\n",
       "      <td>35</td>\n",
       "      <td>...</td>\n",
       "      <td>106</td>\n",
       "      <td>54</td>\n",
       "      <td>0</td>\n",
       "      <td>74</td>\n",
       "      <td>0</td>\n",
       "      <td>77</td>\n",
       "      <td>2</td>\n",
       "      <td>71</td>\n",
       "      <td>37</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>270</td>\n",
       "      <td>294</td>\n",
       "      <td>39</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>177</td>\n",
       "      <td>106</td>\n",
       "      <td>0</td>\n",
       "      <td>321</td>\n",
       "      <td>80</td>\n",
       "      <td>...</td>\n",
       "      <td>356</td>\n",
       "      <td>183</td>\n",
       "      <td>0</td>\n",
       "      <td>231</td>\n",
       "      <td>0</td>\n",
       "      <td>269</td>\n",
       "      <td>0</td>\n",
       "      <td>353</td>\n",
       "      <td>96</td>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>64</td>\n",
       "      <td>61</td>\n",
       "      <td>12</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>49</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>96</td>\n",
       "      <td>16</td>\n",
       "      <td>...</td>\n",
       "      <td>99</td>\n",
       "      <td>34</td>\n",
       "      <td>0</td>\n",
       "      <td>81</td>\n",
       "      <td>0</td>\n",
       "      <td>78</td>\n",
       "      <td>1</td>\n",
       "      <td>78</td>\n",
       "      <td>20</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>114</td>\n",
       "      <td>71</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>65</td>\n",
       "      <td>53</td>\n",
       "      <td>0</td>\n",
       "      <td>157</td>\n",
       "      <td>25</td>\n",
       "      <td>...</td>\n",
       "      <td>148</td>\n",
       "      <td>89</td>\n",
       "      <td>0</td>\n",
       "      <td>108</td>\n",
       "      <td>0</td>\n",
       "      <td>119</td>\n",
       "      <td>1</td>\n",
       "      <td>155</td>\n",
       "      <td>45</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>145</td>\n",
       "      <td>98</td>\n",
       "      <td>25</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>96</td>\n",
       "      <td>92</td>\n",
       "      <td>0</td>\n",
       "      <td>182</td>\n",
       "      <td>50</td>\n",
       "      <td>...</td>\n",
       "      <td>248</td>\n",
       "      <td>79</td>\n",
       "      <td>0</td>\n",
       "      <td>172</td>\n",
       "      <td>0</td>\n",
       "      <td>163</td>\n",
       "      <td>1</td>\n",
       "      <td>198</td>\n",
       "      <td>50</td>\n",
       "      <td>97</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>237 rows × 109 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     TRAV35*01  TRAJ43*01  TRAJ8*01  TRAJ26*01  TRAJ25*01  TRAV8-6*01  \\\n",
       "0          634        615        61         12         12         377   \n",
       "1          361        216       103          9          5         278   \n",
       "2          605        470        82          8          4         382   \n",
       "3          466        382        89          7          7         303   \n",
       "4           32         24         4          0          0          20   \n",
       "..         ...        ...       ...        ...        ...         ...   \n",
       "232         52         28         8          1          1          34   \n",
       "233        270        294        39          1          2         177   \n",
       "234         64         61        12          3          2          49   \n",
       "235        114         71         6          0          0          65   \n",
       "236        145         98        25          1          0          96   \n",
       "\n",
       "     TRAJ27*01  TRAJ19*01  TRAV9-2*01  TRAV1-1*01  ...  TRAV17*01  TRAJ6*01  \\\n",
       "0          269          0         663         178  ...        785       278   \n",
       "1          213          0         479         169  ...        537       233   \n",
       "2          279          0         671         227  ...        835       350   \n",
       "3          213          1         499         162  ...        626       212   \n",
       "4           13          0          35          21  ...         55        16   \n",
       "..         ...        ...         ...         ...  ...        ...       ...   \n",
       "232         40          0          91          35  ...        106        54   \n",
       "233        106          0         321          80  ...        356       183   \n",
       "234         29          0          96          16  ...         99        34   \n",
       "235         53          0         157          25  ...        148        89   \n",
       "236         92          0         182          50  ...        248        79   \n",
       "\n",
       "     TRAJ21*01  TRAV19*01  TRAV18*01  TRAV12-1*01  TRAV9-1*01  TRAV26-1*01  \\\n",
       "0            1        634          0          646           8          632   \n",
       "1            0        546          0          537           4          498   \n",
       "2            1        646          1          685          10          633   \n",
       "3            1        495          0          498           7          462   \n",
       "4            0         16          0           43           1           44   \n",
       "..         ...        ...        ...          ...         ...          ...   \n",
       "232          0         74          0           77           2           71   \n",
       "233          0        231          0          269           0          353   \n",
       "234          0         81          0           78           1           78   \n",
       "235          0        108          0          119           1          155   \n",
       "236          0        172          0          163           1          198   \n",
       "\n",
       "     TRAJ9*01  TRAJ61*01  \n",
       "0         187        369  \n",
       "1         166        180  \n",
       "2         268        306  \n",
       "3         157        291  \n",
       "4          15          7  \n",
       "..        ...        ...  \n",
       "232        37         24  \n",
       "233        96        109  \n",
       "234        20         38  \n",
       "235        45         56  \n",
       "236        50         97  \n",
       "\n",
       "[237 rows x 109 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset.evaluate_segment_usage()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "42a26cc2-325f-497d-9dbb-f911869ac3ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "folder_to_run_mapping = {}\n",
    "for folder in dataset.metadata[['run', 'folder']].folder.unique():\n",
    "    folder_to_run_mapping[folder] = set(dataset.metadata[dataset.metadata.folder == folder].run)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "76439513-adf1-4000-902a-797417ce16d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "norm_usage_table_v = NormalizedSegmentUsageTable.load_from_repertoire_dataset(repertoire_dataset=dataset, gene='TRA', segment_type='V')\n",
    "norm_usage_table_j = NormalizedSegmentUsageTable.load_from_repertoire_dataset(repertoire_dataset=dataset, gene='TRA', segment_type='J')\n",
    "\n",
    "z_score_usage_table_v = StandardizedSegmentUsageTable.load_from_repertoire_dataset(repertoire_dataset=dataset, gene='TRA', segment_type='V', group_mapping=folder_to_run_mapping, standardization_method='log_exp')\n",
    "z_score_usage_table_j = StandardizedSegmentUsageTable.load_from_repertoire_dataset(repertoire_dataset=dataset, gene='TRA', segment_type='J', group_mapping=folder_to_run_mapping, standardization_method='log_exp')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "f1a4189f-c558-448a-9d39-3174bf17de5b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "     TRAV35*01  TRAV8-6*01  TRAV9-2*01  TRAV1-1*01  TRAV22*01  TRAV36DV7*01  \\\n",
       "0     0.035115    0.020881    0.036721    0.009859   0.019662      0.017945   \n",
       "1     0.026544    0.020441    0.035221    0.012426   0.018971      0.013529   \n",
       "2     0.031757    0.020051    0.035221    0.011915   0.021416      0.015485   \n",
       "3     0.032138    0.020897    0.034414    0.011172   0.022966      0.014414   \n",
       "4     0.030740    0.019212    0.033622    0.020173   0.024015      0.016330   \n",
       "..         ...         ...         ...         ...        ...           ...   \n",
       "232   0.024857    0.016252    0.043499    0.016730   0.016252      0.009082   \n",
       "233   0.033826    0.022175    0.040215    0.010023   0.020296      0.013781   \n",
       "234   0.028046    0.021472    0.042068    0.007011   0.015337      0.013146   \n",
       "235   0.033918    0.019339    0.046712    0.007438   0.019637      0.002380   \n",
       "236   0.028844    0.019097    0.036204    0.009946   0.019296      0.018301   \n",
       "\n",
       "     TRAV13-1*01  TRAV6*01  TRAV8-4*01  TRAV21*01  ...  TRAV12-2*01  \\\n",
       "0       0.041983  0.016782    0.018721   0.037497  ...     0.022985   \n",
       "1       0.042868  0.019191    0.020956   0.040000  ...     0.025221   \n",
       "2       0.042570  0.017532    0.019946   0.039001  ...     0.025773   \n",
       "3       0.045034  0.018000    0.018966   0.039862  ...     0.025034   \n",
       "4       0.050913  0.018252    0.015370   0.025937  ...     0.025937   \n",
       "..           ...       ...         ...        ...  ...          ...   \n",
       "232     0.054493  0.015774    0.022945   0.029159  ...     0.029637   \n",
       "233     0.048359  0.014282    0.019293   0.032197  ...     0.026309   \n",
       "234     0.047327  0.023663    0.018405   0.040316  ...     0.020596   \n",
       "235     0.044927  0.016959    0.010711   0.045225  ...     0.027670   \n",
       "236     0.044957  0.015715    0.019893   0.039387  ...     0.027253   \n",
       "\n",
       "     TRAV25*01  TRAV26-2*01  TRAV2*01  TRAV17*01  TRAV19*01  TRAV18*01  \\\n",
       "0     0.021988     0.035060  0.015453   0.043478   0.035115   0.000000   \n",
       "1     0.017868     0.030368  0.020662   0.039485   0.040147   0.000000   \n",
       "2     0.020629     0.033122  0.020524   0.043830   0.033909   0.000052   \n",
       "3     0.017862     0.031724  0.019034   0.043172   0.034138   0.000000   \n",
       "4     0.026897     0.018252  0.024976   0.052834   0.015370   0.000000   \n",
       "..         ...          ...       ...        ...        ...        ...   \n",
       "232   0.026291     0.025813  0.021511   0.050669   0.035373   0.000000   \n",
       "233   0.019920     0.031696  0.016287   0.044600   0.028940   0.000000   \n",
       "234   0.018843     0.022787  0.026293   0.043383   0.035495   0.000000   \n",
       "235   0.022315     0.033323  0.017554   0.044035   0.032133   0.000000   \n",
       "236   0.021484     0.031430  0.017903   0.049334   0.034215   0.000000   \n",
       "\n",
       "     TRAV12-1*01  TRAV9-1*01  TRAV26-1*01  \n",
       "0       0.035780    0.000443     0.035004  \n",
       "1       0.039485    0.000294     0.036618  \n",
       "2       0.035956    0.000525     0.033227  \n",
       "3       0.034345    0.000483     0.031862  \n",
       "4       0.041306    0.000961     0.042267  \n",
       "..           ...         ...          ...  \n",
       "232     0.036807    0.000956     0.033939  \n",
       "233     0.033701    0.000000     0.044225  \n",
       "234     0.034181    0.000438     0.034181  \n",
       "235     0.035406    0.000298     0.046117  \n",
       "236     0.032425    0.000199     0.039387  \n",
       "\n",
       "[237 rows x 48 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "norm_usage_table_v"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a7827db-1f25-4b60-b3b8-d9b7c0c1bf83",
   "metadata": {},
   "source": [
    "#### Sampling reads from the repertoire, not changing the gene usage distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "ced538a7-622c-4ba7-8c69-0f417b89fc05",
   "metadata": {},
   "outputs": [],
   "source": [
    "sampled_repertoire = RepertoireSampling().sample(repertoire=dataset[0], old_usage_matrix=[norm_usage_table_v, norm_usage_table_j], new_usage_matrix=None, n=100000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "f89df81c-f059-4d98-a7a2-38a8c0e13378",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Repertoire of 12315 clonotypes and 100000 cells:\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CAVRDG*_QGGKLIF cdr3nt_0\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CVVSPLW_NNRLAF cdr3nt_2\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CATGPTR_GTALIF cdr3nt_4\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CAGVPRT_GNQFYF cdr3nt_5\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CIVRLVP_VNRKLTF cdr3nt_6\n",
       "file.id_short                                                                20000200808\n",
       "sample.age                                                                            19\n",
       "sample.sex                                                                          male\n",
       "sample.height(cm)                                                                  187.0\n",
       "sample.weight(kg)                                                                  105.0\n",
       "sample.waist(cm)                                                                    97.0\n",
       "sample.diastolicBloodPressure(mmHg)                                                   80\n",
       "sample.bodyTemperature                                                              36.6\n",
       "sample.pulse(beats/min)                                                               74\n",
       "sample.oxygenLevel(%)                                                               99.0\n",
       "sample.erythrocytes(10*12/l)                                                        5.03\n",
       "sample.hemoglobinCount(g/l)                                                        172.0\n",
       "sample.hematocrit(%)                                                                50.0\n",
       "sample.plateletCount(10*9/l)                                                       256.0\n",
       "sample.reticulocyteCount(10*9/l)                                                     NaN\n",
       "sample.WBC(10*9/l)                                                                  10.7\n",
       "sample.ESR(mm/h)                                                                     4.0\n",
       "sample.serumTotalProtein(g/l)                                                       78.8\n",
       "sample.hemoglobin(g/l)                                                               NaN\n",
       "sample.COVID_PCR                                                                negative\n",
       "sample.bloodType                                                                  B(III)\n",
       "sample.RhFactor                                                                 poditive\n",
       "sample.COVID_IgM                                                                      no\n",
       "sample.COVID_IgG                                                                      no\n",
       "sample.COVID_status                                                              healthy\n",
       "sample.neutrophil(10*9/l)                                                            7.2\n",
       "sample.lymphocytes(10*9/l)                                                           2.7\n",
       "sample.monocyte(10*9/l)                                                              0.8\n",
       "sample.eosinophils(10*9/l)                                                           NaN\n",
       "sample.basophil(10*9/l)                                                              NaN\n",
       "sample.NLR                                                                           NaN\n",
       "sample.neutrophilFraction                                                           69.0\n",
       "sample.lymphocytesFraction                                                          24.0\n",
       "sample.monocyteFraction                                                              6.0\n",
       "sample.eosinophilsFraction                                                           1.0\n",
       "sample.basophilFraction                                                              0.0\n",
       "sample.HLA-A.1                                                                   A*02:01\n",
       "sample.HLA-A.2                                                                   A*11:01\n",
       "sample.HLA-B.1                                                                   B*40:01\n",
       "sample.HLA-B.2                                                                   B*52:01\n",
       "sample.HLA-C.1                                                                   C*03:04\n",
       "sample.HLA-C.2                                                                   C*12:02\n",
       "sample.HLA-DPB1.1                                                             DPB1*04:01\n",
       "sample.HLA-DPB1.2                                                             DPB1*06:01\n",
       "sample.HLA-DQB1.1                                                             DQB1*03:02\n",
       "sample.HLA-DQB1.2                                                             DQB1*05:01\n",
       "sample.HLA-DRB1.1                                                             DRB1*01:01\n",
       "sample.HLA-DRB1.2                                                             DRB1*04:04\n",
       "Unnamed: 48                                                                          NaN\n",
       "Unnamed: 49                                                                          NaN\n",
       "Unnamed: 50                                                                          NaN\n",
       "Unnamed: 51                                                                          NaN\n",
       "Unnamed: 52                                                                          NaN\n",
       "Unnamed: 53                                                                          NaN\n",
       "run                                            020000200808_S181_L002.clonotypes.TRA.txt\n",
       "folder                                                             2020/10_FMBA_NovaSeq5\n",
       "path                                   ../../../data/cohort_I/nonfunctional/020000200...\n",
       "Name: 0, dtype: object\n",
       "..."
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampled_repertoire"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb9a7cca-1247-44ba-a4a4-e0fa7028e4e9",
   "metadata": {},
   "source": [
    "#### Resampling the repertoire using the new derived usages for V and J genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "2e821c34-8586-4ff9-b199-0177d368fe85",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Repertoire of 18055 clonotypes and 4514408 cells:\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CAVRDG*_QGGKLIF cdr3nt_0\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CVVSPLW_NNRLAF cdr3nt_2\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CATGPTR_GTALIF cdr3nt_4\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CAGVPRT_GNQFYF cdr3nt_5\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CIVRLVP_VNRKLTF cdr3nt_6\n",
       "file.id_short                                                                20000200808\n",
       "sample.age                                                                            19\n",
       "sample.sex                                                                          male\n",
       "sample.height(cm)                                                                  187.0\n",
       "sample.weight(kg)                                                                  105.0\n",
       "sample.waist(cm)                                                                    97.0\n",
       "sample.diastolicBloodPressure(mmHg)                                                   80\n",
       "sample.bodyTemperature                                                              36.6\n",
       "sample.pulse(beats/min)                                                               74\n",
       "sample.oxygenLevel(%)                                                               99.0\n",
       "sample.erythrocytes(10*12/l)                                                        5.03\n",
       "sample.hemoglobinCount(g/l)                                                        172.0\n",
       "sample.hematocrit(%)                                                                50.0\n",
       "sample.plateletCount(10*9/l)                                                       256.0\n",
       "sample.reticulocyteCount(10*9/l)                                                     NaN\n",
       "sample.WBC(10*9/l)                                                                  10.7\n",
       "sample.ESR(mm/h)                                                                     4.0\n",
       "sample.serumTotalProtein(g/l)                                                       78.8\n",
       "sample.hemoglobin(g/l)                                                               NaN\n",
       "sample.COVID_PCR                                                                negative\n",
       "sample.bloodType                                                                  B(III)\n",
       "sample.RhFactor                                                                 poditive\n",
       "sample.COVID_IgM                                                                      no\n",
       "sample.COVID_IgG                                                                      no\n",
       "sample.COVID_status                                                              healthy\n",
       "sample.neutrophil(10*9/l)                                                            7.2\n",
       "sample.lymphocytes(10*9/l)                                                           2.7\n",
       "sample.monocyte(10*9/l)                                                              0.8\n",
       "sample.eosinophils(10*9/l)                                                           NaN\n",
       "sample.basophil(10*9/l)                                                              NaN\n",
       "sample.NLR                                                                           NaN\n",
       "sample.neutrophilFraction                                                           69.0\n",
       "sample.lymphocytesFraction                                                          24.0\n",
       "sample.monocyteFraction                                                              6.0\n",
       "sample.eosinophilsFraction                                                           1.0\n",
       "sample.basophilFraction                                                              0.0\n",
       "sample.HLA-A.1                                                                   A*02:01\n",
       "sample.HLA-A.2                                                                   A*11:01\n",
       "sample.HLA-B.1                                                                   B*40:01\n",
       "sample.HLA-B.2                                                                   B*52:01\n",
       "sample.HLA-C.1                                                                   C*03:04\n",
       "sample.HLA-C.2                                                                   C*12:02\n",
       "sample.HLA-DPB1.1                                                             DPB1*04:01\n",
       "sample.HLA-DPB1.2                                                             DPB1*06:01\n",
       "sample.HLA-DQB1.1                                                             DQB1*03:02\n",
       "sample.HLA-DQB1.2                                                             DQB1*05:01\n",
       "sample.HLA-DRB1.1                                                             DRB1*01:01\n",
       "sample.HLA-DRB1.2                                                             DRB1*04:04\n",
       "Unnamed: 48                                                                          NaN\n",
       "Unnamed: 49                                                                          NaN\n",
       "Unnamed: 50                                                                          NaN\n",
       "Unnamed: 51                                                                          NaN\n",
       "Unnamed: 52                                                                          NaN\n",
       "Unnamed: 53                                                                          NaN\n",
       "run                                            020000200808_S181_L002.clonotypes.TRA.txt\n",
       "folder                                                             2020/10_FMBA_NovaSeq5\n",
       "path                                   ../../../data/cohort_I/nonfunctional/020000200...\n",
       "Name: 0, dtype: object\n",
       "..."
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "446e7f43-fb73-4119-a46f-332e7aaa0308",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.8143439292907715\n"
     ]
    }
   ],
   "source": [
    "t0 = time.time()\n",
    "resampled_repertoire = RepertoireSampling().sample(repertoire=dataset[0], old_usage_matrix=[norm_usage_table_v, norm_usage_table_j], new_usage_matrix=[z_score_usage_table_v, z_score_usage_table_j])\n",
    "print(time.time() - t0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "758e6e25-bfdc-4e22-8af3-124296a8b897",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Repertoire of 17238 clonotypes and 4514408 cells:\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CAVRDG*_QGGKLIF cdr3nt_0\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CVVSPLW_NNRLAF cdr3nt_2\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CATGPTR_GTALIF cdr3nt_4\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CAVEW_QKLLF cdr3nt_8\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CAFRMM_GFKTIF cdr3nt_14\n",
       "file.id_short                                                                20000200808\n",
       "sample.age                                                                            19\n",
       "sample.sex                                                                          male\n",
       "sample.height(cm)                                                                  187.0\n",
       "sample.weight(kg)                                                                  105.0\n",
       "sample.waist(cm)                                                                    97.0\n",
       "sample.diastolicBloodPressure(mmHg)                                                   80\n",
       "sample.bodyTemperature                                                              36.6\n",
       "sample.pulse(beats/min)                                                               74\n",
       "sample.oxygenLevel(%)                                                               99.0\n",
       "sample.erythrocytes(10*12/l)                                                        5.03\n",
       "sample.hemoglobinCount(g/l)                                                        172.0\n",
       "sample.hematocrit(%)                                                                50.0\n",
       "sample.plateletCount(10*9/l)                                                       256.0\n",
       "sample.reticulocyteCount(10*9/l)                                                     NaN\n",
       "sample.WBC(10*9/l)                                                                  10.7\n",
       "sample.ESR(mm/h)                                                                     4.0\n",
       "sample.serumTotalProtein(g/l)                                                       78.8\n",
       "sample.hemoglobin(g/l)                                                               NaN\n",
       "sample.COVID_PCR                                                                negative\n",
       "sample.bloodType                                                                  B(III)\n",
       "sample.RhFactor                                                                 poditive\n",
       "sample.COVID_IgM                                                                      no\n",
       "sample.COVID_IgG                                                                      no\n",
       "sample.COVID_status                                                              healthy\n",
       "sample.neutrophil(10*9/l)                                                            7.2\n",
       "sample.lymphocytes(10*9/l)                                                           2.7\n",
       "sample.monocyte(10*9/l)                                                              0.8\n",
       "sample.eosinophils(10*9/l)                                                           NaN\n",
       "sample.basophil(10*9/l)                                                              NaN\n",
       "sample.NLR                                                                           NaN\n",
       "sample.neutrophilFraction                                                           69.0\n",
       "sample.lymphocytesFraction                                                          24.0\n",
       "sample.monocyteFraction                                                              6.0\n",
       "sample.eosinophilsFraction                                                           1.0\n",
       "sample.basophilFraction                                                              0.0\n",
       "sample.HLA-A.1                                                                   A*02:01\n",
       "sample.HLA-A.2                                                                   A*11:01\n",
       "sample.HLA-B.1                                                                   B*40:01\n",
       "sample.HLA-B.2                                                                   B*52:01\n",
       "sample.HLA-C.1                                                                   C*03:04\n",
       "sample.HLA-C.2                                                                   C*12:02\n",
       "sample.HLA-DPB1.1                                                             DPB1*04:01\n",
       "sample.HLA-DPB1.2                                                             DPB1*06:01\n",
       "sample.HLA-DQB1.1                                                             DQB1*03:02\n",
       "sample.HLA-DQB1.2                                                             DQB1*05:01\n",
       "sample.HLA-DRB1.1                                                             DRB1*01:01\n",
       "sample.HLA-DRB1.2                                                             DRB1*04:04\n",
       "Unnamed: 48                                                                          NaN\n",
       "Unnamed: 49                                                                          NaN\n",
       "Unnamed: 50                                                                          NaN\n",
       "Unnamed: 51                                                                          NaN\n",
       "Unnamed: 52                                                                          NaN\n",
       "Unnamed: 53                                                                          NaN\n",
       "run                                            020000200808_S181_L002.clonotypes.TRA.txt\n",
       "folder                                                             2020/10_FMBA_NovaSeq5\n",
       "path                                   ../../../data/cohort_I/nonfunctional/020000200...\n",
       "Name: 0, dtype: object\n",
       "..."
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resampled_repertoire"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66465e2d-d58b-48ba-bb98-e56281365c45",
   "metadata": {},
   "source": [
    "#### Resampling all the reperoires from a dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "d9096b41-f8d2-4489-8190-437d8196bc59",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "248.3766894340515\n"
     ]
    }
   ],
   "source": [
    "t0 = time.time()\n",
    "resampled_dataset = dataset.resample(updated_segment_usage_tables=[z_score_usage_table_v, z_score_usage_table_j], threads=28)\n",
    "print(time.time() - t0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "7fdbddd3-271c-448f-bc9e-9f9c73988d47",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Repertoire of 15997 clonotypes and 3186951 cells:\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CAGLKRG_AQKLVF cdr3nt_8\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CASQICS_GNNRLAF cdr3nt_9\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CIVIRLCG_TGNQFYF cdr3nt_10\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CIPL_NMRF cdr3nt_12\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CAVSPCM_NNRLAF cdr3nt_14\n",
       "file.id_short                                                                50001340808\n",
       "sample.age                                                                            28\n",
       "sample.sex                                                                        female\n",
       "sample.height(cm)                                                                  160.0\n",
       "sample.weight(kg)                                                                   64.0\n",
       "sample.waist(cm)                                                                    67.0\n",
       "sample.diastolicBloodPressure(mmHg)                                                   60\n",
       "sample.bodyTemperature                                                              36.3\n",
       "sample.pulse(beats/min)                                                               64\n",
       "sample.oxygenLevel(%)                                                               98.0\n",
       "sample.erythrocytes(10*12/l)                                                        4.78\n",
       "sample.hemoglobinCount(g/l)                                                        116.0\n",
       "sample.hematocrit(%)                                                                36.9\n",
       "sample.plateletCount(10*9/l)                                                       386.0\n",
       "sample.reticulocyteCount(10*9/l)                                                   0.357\n",
       "sample.WBC(10*9/l)                                                                  4.24\n",
       "sample.ESR(mm/h)                                                                    11.0\n",
       "sample.serumTotalProtein(g/l)                                                       75.0\n",
       "sample.hemoglobin(g/l)                                                               NaN\n",
       "sample.COVID_PCR                                                                negative\n",
       "sample.bloodType                                                                   A(II)\n",
       "sample.RhFactor                                                                 poditive\n",
       "sample.COVID_IgM                                                                      no\n",
       "sample.COVID_IgG                                                                      no\n",
       "sample.COVID_status                                                              healthy\n",
       "sample.neutrophil(10*9/l)                                                            NaN\n",
       "sample.lymphocytes(10*9/l)                                                           NaN\n",
       "sample.monocyte(10*9/l)                                                              NaN\n",
       "sample.eosinophils(10*9/l)                                                           NaN\n",
       "sample.basophil(10*9/l)                                                              NaN\n",
       "sample.NLR                                                                           NaN\n",
       "sample.neutrophilFraction                                                            NaN\n",
       "sample.lymphocytesFraction                                                          31.5\n",
       "sample.monocyteFraction                                                              1.0\n",
       "sample.eosinophilsFraction                                                           NaN\n",
       "sample.basophilFraction                                                              NaN\n",
       "sample.HLA-A.1                                                                   A*02:01\n",
       "sample.HLA-A.2                                                                   A*32:01\n",
       "sample.HLA-B.1                                                                   B*07:02\n",
       "sample.HLA-B.2                                                                   B*37:01\n",
       "sample.HLA-C.1                                                                   C*02:02\n",
       "sample.HLA-C.2                                                                   C*07:02\n",
       "sample.HLA-DPB1.1                                                             DPB1*03:01\n",
       "sample.HLA-DPB1.2                                                             DPB1*04:01\n",
       "sample.HLA-DQB1.1                                                             DQB1*03:01\n",
       "sample.HLA-DQB1.2                                                             DQB1*06:02\n",
       "sample.HLA-DRB1.1                                                             DRB1*08:03\n",
       "sample.HLA-DRB1.2                                                             DRB1*15:01\n",
       "Unnamed: 48                                                                          NaN\n",
       "Unnamed: 49                                                                          NaN\n",
       "Unnamed: 50                                                                          NaN\n",
       "Unnamed: 51                                                                          NaN\n",
       "Unnamed: 52                                                                          NaN\n",
       "Unnamed: 53                                                                          NaN\n",
       "run                                             050001340808_S85_L001.clonotypes.TRA.txt\n",
       "folder                                                         2020/12_FMBA_Novaseq7_DNA\n",
       "path                                   ../../../data/cohort_I/nonfunctional/050001340...\n",
       "Name: 14, dtype: object\n",
       "..."
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset[14]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "e2c26d1f-6010-4617-8d27-38c8fe4bb2d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Repertoire of 15466 clonotypes and 3186951 cells:\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CAVLQLQQIFQ*VGNFNKFYF cdr3nt_3958\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CIVIRLCG_TGNQFYF cdr3nt_10\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CAT*L_NKFYF cdr3nt_2072\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CAVDTT_NKFYF cdr3nt_324\n",
       "κIndex(['count', 'freq', 'cdr3nt', 'cdr3aa', 'v', 'd', 'j', 'VEnd', 'DStart',\n",
       "       'DEnd', 'JStart'],\n",
       "      dtype='object') CVVSAS_DKFYF cdr3nt_410\n",
       "file.id_short                                                                50001340808\n",
       "sample.age                                                                            28\n",
       "sample.sex                                                                        female\n",
       "sample.height(cm)                                                                  160.0\n",
       "sample.weight(kg)                                                                   64.0\n",
       "sample.waist(cm)                                                                    67.0\n",
       "sample.diastolicBloodPressure(mmHg)                                                   60\n",
       "sample.bodyTemperature                                                              36.3\n",
       "sample.pulse(beats/min)                                                               64\n",
       "sample.oxygenLevel(%)                                                               98.0\n",
       "sample.erythrocytes(10*12/l)                                                        4.78\n",
       "sample.hemoglobinCount(g/l)                                                        116.0\n",
       "sample.hematocrit(%)                                                                36.9\n",
       "sample.plateletCount(10*9/l)                                                       386.0\n",
       "sample.reticulocyteCount(10*9/l)                                                   0.357\n",
       "sample.WBC(10*9/l)                                                                  4.24\n",
       "sample.ESR(mm/h)                                                                    11.0\n",
       "sample.serumTotalProtein(g/l)                                                       75.0\n",
       "sample.hemoglobin(g/l)                                                               NaN\n",
       "sample.COVID_PCR                                                                negative\n",
       "sample.bloodType                                                                   A(II)\n",
       "sample.RhFactor                                                                 poditive\n",
       "sample.COVID_IgM                                                                      no\n",
       "sample.COVID_IgG                                                                      no\n",
       "sample.COVID_status                                                              healthy\n",
       "sample.neutrophil(10*9/l)                                                            NaN\n",
       "sample.lymphocytes(10*9/l)                                                           NaN\n",
       "sample.monocyte(10*9/l)                                                              NaN\n",
       "sample.eosinophils(10*9/l)                                                           NaN\n",
       "sample.basophil(10*9/l)                                                              NaN\n",
       "sample.NLR                                                                           NaN\n",
       "sample.neutrophilFraction                                                            NaN\n",
       "sample.lymphocytesFraction                                                          31.5\n",
       "sample.monocyteFraction                                                              1.0\n",
       "sample.eosinophilsFraction                                                           NaN\n",
       "sample.basophilFraction                                                              NaN\n",
       "sample.HLA-A.1                                                                   A*02:01\n",
       "sample.HLA-A.2                                                                   A*32:01\n",
       "sample.HLA-B.1                                                                   B*07:02\n",
       "sample.HLA-B.2                                                                   B*37:01\n",
       "sample.HLA-C.1                                                                   C*02:02\n",
       "sample.HLA-C.2                                                                   C*07:02\n",
       "sample.HLA-DPB1.1                                                             DPB1*03:01\n",
       "sample.HLA-DPB1.2                                                             DPB1*04:01\n",
       "sample.HLA-DQB1.1                                                             DQB1*03:01\n",
       "sample.HLA-DQB1.2                                                             DQB1*06:02\n",
       "sample.HLA-DRB1.1                                                             DRB1*08:03\n",
       "sample.HLA-DRB1.2                                                             DRB1*15:01\n",
       "Unnamed: 48                                                                          NaN\n",
       "Unnamed: 49                                                                          NaN\n",
       "Unnamed: 50                                                                          NaN\n",
       "Unnamed: 51                                                                          NaN\n",
       "Unnamed: 52                                                                          NaN\n",
       "Unnamed: 53                                                                          NaN\n",
       "run                                             050001340808_S85_L001.clonotypes.TRA.txt\n",
       "folder                                                         2020/12_FMBA_Novaseq7_DNA\n",
       "path                                   ../../../data/cohort_I/nonfunctional/050001340...\n",
       "Name: 14, dtype: object\n",
       "..."
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resampled_dataset[14]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "48041979-dd83-45be-9a91-5474144f9b15",
   "metadata": {},
   "outputs": [],
   "source": [
    "nf_trav_matrix = NormalizedSegmentUsageTable.load_from_repertoire_dataset(\n",
    "    resampled_dataset, gene='TRA', segment_type='V').segment_usage_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "9899f213-125f-4ade-995e-75d151ff2734",
   "metadata": {},
   "outputs": [],
   "source": [
    "nf_traj_matrix = NormalizedSegmentUsageTable.load_from_repertoire_dataset(\n",
    "    resampled_dataset, gene='TRA', segment_type='J').segment_usage_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "e3f2f91b-d55d-4782-bd5c-476582d53beb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TRAJ43*01</th>\n",
       "      <th>TRAJ8*01</th>\n",
       "      <th>TRAJ26*01</th>\n",
       "      <th>TRAJ25*01</th>\n",
       "      <th>TRAJ27*01</th>\n",
       "      <th>TRAJ19*01</th>\n",
       "      <th>TRAJ29*01</th>\n",
       "      <th>TRAJ28*01</th>\n",
       "      <th>TRAJ38*01</th>\n",
       "      <th>TRAJ18*01</th>\n",
       "      <th>...</th>\n",
       "      <th>TRAJ14*01</th>\n",
       "      <th>TRAJ40*01</th>\n",
       "      <th>TRAJ15*01</th>\n",
       "      <th>TRAJ41*01</th>\n",
       "      <th>TRAJ13*01</th>\n",
       "      <th>TRAJ57*01</th>\n",
       "      <th>TRAJ6*01</th>\n",
       "      <th>TRAJ21*01</th>\n",
       "      <th>TRAJ9*01</th>\n",
       "      <th>TRAJ61*01</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.033821</td>\n",
       "      <td>0.003307</td>\n",
       "      <td>0.000638</td>\n",
       "      <td>0.000464</td>\n",
       "      <td>0.015257</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.023611</td>\n",
       "      <td>0.025525</td>\n",
       "      <td>0.013343</td>\n",
       "      <td>0.008992</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.028310</td>\n",
       "      <td>0.015199</td>\n",
       "      <td>0.015431</td>\n",
       "      <td>0.013807</td>\n",
       "      <td>0.024249</td>\n",
       "      <td>0.015605</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.010790</td>\n",
       "      <td>0.020362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.016863</td>\n",
       "      <td>0.007651</td>\n",
       "      <td>0.000703</td>\n",
       "      <td>0.000156</td>\n",
       "      <td>0.015926</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.028339</td>\n",
       "      <td>0.023187</td>\n",
       "      <td>0.010227</td>\n",
       "      <td>0.008666</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000468</td>\n",
       "      <td>0.022953</td>\n",
       "      <td>0.014521</td>\n",
       "      <td>0.013740</td>\n",
       "      <td>0.015380</td>\n",
       "      <td>0.022875</td>\n",
       "      <td>0.017332</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.012491</td>\n",
       "      <td>0.013428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.025352</td>\n",
       "      <td>0.004253</td>\n",
       "      <td>0.000331</td>\n",
       "      <td>0.000166</td>\n",
       "      <td>0.014637</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.030544</td>\n",
       "      <td>0.023364</td>\n",
       "      <td>0.010052</td>\n",
       "      <td>0.008893</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.029716</td>\n",
       "      <td>0.016073</td>\n",
       "      <td>0.016570</td>\n",
       "      <td>0.017012</td>\n",
       "      <td>0.010218</td>\n",
       "      <td>0.018835</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.014195</td>\n",
       "      <td>0.015742</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.027033</td>\n",
       "      <td>0.005967</td>\n",
       "      <td>0.000431</td>\n",
       "      <td>0.000216</td>\n",
       "      <td>0.014954</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.026026</td>\n",
       "      <td>0.025883</td>\n",
       "      <td>0.013804</td>\n",
       "      <td>0.010497</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000144</td>\n",
       "      <td>0.026242</td>\n",
       "      <td>0.014092</td>\n",
       "      <td>0.015601</td>\n",
       "      <td>0.015242</td>\n",
       "      <td>0.014451</td>\n",
       "      <td>0.014954</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.010928</td>\n",
       "      <td>0.019915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.024803</td>\n",
       "      <td>0.004510</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013529</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.024803</td>\n",
       "      <td>0.041714</td>\n",
       "      <td>0.012401</td>\n",
       "      <td>0.009019</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.019166</td>\n",
       "      <td>0.027057</td>\n",
       "      <td>0.011274</td>\n",
       "      <td>0.021421</td>\n",
       "      <td>0.001127</td>\n",
       "      <td>0.016911</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.014656</td>\n",
       "      <td>0.007892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>0.013804</td>\n",
       "      <td>0.003579</td>\n",
       "      <td>0.000511</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.019427</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.037832</td>\n",
       "      <td>0.027607</td>\n",
       "      <td>0.007669</td>\n",
       "      <td>0.012270</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.018916</td>\n",
       "      <td>0.016360</td>\n",
       "      <td>0.016360</td>\n",
       "      <td>0.021472</td>\n",
       "      <td>0.004601</td>\n",
       "      <td>0.025562</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.018916</td>\n",
       "      <td>0.012270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>0.037503</td>\n",
       "      <td>0.005009</td>\n",
       "      <td>0.000128</td>\n",
       "      <td>0.000257</td>\n",
       "      <td>0.013486</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.028127</td>\n",
       "      <td>0.022219</td>\n",
       "      <td>0.011816</td>\n",
       "      <td>0.009761</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.029926</td>\n",
       "      <td>0.014385</td>\n",
       "      <td>0.020293</td>\n",
       "      <td>0.018366</td>\n",
       "      <td>0.015155</td>\n",
       "      <td>0.022990</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.012073</td>\n",
       "      <td>0.013743</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>0.027099</td>\n",
       "      <td>0.005331</td>\n",
       "      <td>0.001333</td>\n",
       "      <td>0.000888</td>\n",
       "      <td>0.011995</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.020435</td>\n",
       "      <td>0.025766</td>\n",
       "      <td>0.019991</td>\n",
       "      <td>0.009329</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000444</td>\n",
       "      <td>0.025766</td>\n",
       "      <td>0.020435</td>\n",
       "      <td>0.010662</td>\n",
       "      <td>0.016881</td>\n",
       "      <td>0.022657</td>\n",
       "      <td>0.015104</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.008885</td>\n",
       "      <td>0.016437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>0.022486</td>\n",
       "      <td>0.001606</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.015740</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021844</td>\n",
       "      <td>0.033730</td>\n",
       "      <td>0.003855</td>\n",
       "      <td>0.011564</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.020238</td>\n",
       "      <td>0.019274</td>\n",
       "      <td>0.018632</td>\n",
       "      <td>0.017347</td>\n",
       "      <td>0.008352</td>\n",
       "      <td>0.025699</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.014134</td>\n",
       "      <td>0.017668</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>0.019531</td>\n",
       "      <td>0.004112</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.018914</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.029194</td>\n",
       "      <td>0.020970</td>\n",
       "      <td>0.008840</td>\n",
       "      <td>0.008635</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.024877</td>\n",
       "      <td>0.009663</td>\n",
       "      <td>0.012952</td>\n",
       "      <td>0.013980</td>\n",
       "      <td>0.014803</td>\n",
       "      <td>0.014803</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.009457</td>\n",
       "      <td>0.019737</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>237 rows × 61 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     TRAJ43*01  TRAJ8*01  TRAJ26*01  TRAJ25*01  TRAJ27*01  TRAJ19*01  \\\n",
       "0     0.033821  0.003307   0.000638   0.000464   0.015257   0.000000   \n",
       "1     0.016863  0.007651   0.000703   0.000156   0.015926   0.000000   \n",
       "2     0.025352  0.004253   0.000331   0.000166   0.014637   0.000000   \n",
       "3     0.027033  0.005967   0.000431   0.000216   0.014954   0.000072   \n",
       "4     0.024803  0.004510   0.000000   0.000000   0.013529   0.000000   \n",
       "..         ...       ...        ...        ...        ...        ...   \n",
       "232   0.013804  0.003579   0.000511   0.000000   0.019427   0.000000   \n",
       "233   0.037503  0.005009   0.000128   0.000257   0.013486   0.000000   \n",
       "234   0.027099  0.005331   0.001333   0.000888   0.011995   0.000000   \n",
       "235   0.022486  0.001606   0.000000   0.000000   0.015740   0.000000   \n",
       "236   0.019531  0.004112   0.000000   0.000000   0.018914   0.000000   \n",
       "\n",
       "     TRAJ29*01  TRAJ28*01  TRAJ38*01  TRAJ18*01  ...  TRAJ14*01  TRAJ40*01  \\\n",
       "0     0.023611   0.025525   0.013343   0.008992  ...   0.000000   0.028310   \n",
       "1     0.028339   0.023187   0.010227   0.008666  ...   0.000468   0.022953   \n",
       "2     0.030544   0.023364   0.010052   0.008893  ...   0.000000   0.029716   \n",
       "3     0.026026   0.025883   0.013804   0.010497  ...   0.000144   0.026242   \n",
       "4     0.024803   0.041714   0.012401   0.009019  ...   0.000000   0.019166   \n",
       "..         ...        ...        ...        ...  ...        ...        ...   \n",
       "232   0.037832   0.027607   0.007669   0.012270  ...   0.000000   0.018916   \n",
       "233   0.028127   0.022219   0.011816   0.009761  ...   0.000000   0.029926   \n",
       "234   0.020435   0.025766   0.019991   0.009329  ...   0.000444   0.025766   \n",
       "235   0.021844   0.033730   0.003855   0.011564  ...   0.000000   0.020238   \n",
       "236   0.029194   0.020970   0.008840   0.008635  ...   0.000000   0.024877   \n",
       "\n",
       "     TRAJ15*01  TRAJ41*01  TRAJ13*01  TRAJ57*01  TRAJ6*01  TRAJ21*01  \\\n",
       "0     0.015199   0.015431   0.013807   0.024249  0.015605        0.0   \n",
       "1     0.014521   0.013740   0.015380   0.022875  0.017332        0.0   \n",
       "2     0.016073   0.016570   0.017012   0.010218  0.018835        0.0   \n",
       "3     0.014092   0.015601   0.015242   0.014451  0.014954        0.0   \n",
       "4     0.027057   0.011274   0.021421   0.001127  0.016911        0.0   \n",
       "..         ...        ...        ...        ...       ...        ...   \n",
       "232   0.016360   0.016360   0.021472   0.004601  0.025562        0.0   \n",
       "233   0.014385   0.020293   0.018366   0.015155  0.022990        0.0   \n",
       "234   0.020435   0.010662   0.016881   0.022657  0.015104        0.0   \n",
       "235   0.019274   0.018632   0.017347   0.008352  0.025699        0.0   \n",
       "236   0.009663   0.012952   0.013980   0.014803  0.014803        0.0   \n",
       "\n",
       "     TRAJ9*01  TRAJ61*01  \n",
       "0    0.010790   0.020362  \n",
       "1    0.012491   0.013428  \n",
       "2    0.014195   0.015742  \n",
       "3    0.010928   0.019915  \n",
       "4    0.014656   0.007892  \n",
       "..        ...        ...  \n",
       "232  0.018916   0.012270  \n",
       "233  0.012073   0.013743  \n",
       "234  0.008885   0.016437  \n",
       "235  0.014134   0.017668  \n",
       "236  0.009457   0.019737  \n",
       "\n",
       "[237 rows x 61 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nf_traj_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "3df33cf4-95a8-4b50-930c-0501850a9b6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "nf_trav_matrix.to_csv('../../../data/cohort_I/nonfunctional_corrected/nf_v_clonotypes_TRA.csv', index = False)\n",
    "nf_traj_matrix.to_csv('../../../data/cohort_I/nonfunctional_corrected/nf_j_clonotypes_TRA.csv', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf02c951-292c-4789-934c-7c22c3733510",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "58038852-2d4c-4d68-9b26-00edc21c1423",
   "metadata": {},
   "source": [
    "### Onlyfunctional sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d3b7462-8b0c-4cc3-b820-ccfe3b4111f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3574192\n"
     ]
    }
   ],
   "source": [
    "base_path = '../../../data/cohort_I/onlyfunctional/'\n",
    "t0 = time.time()\n",
    "dataset = RepertoireDataset.load(parser=VDJtoolsParser(sep=','), \n",
    "                                 metadata=new_meta,\n",
    "                                 threads=28,\n",
    "                                 gene = 'TRA',\n",
    "                                 paths=[base_path + r['run']\n",
    "                                        for _, r in new_meta.iterrows()])\n",
    "print(time.time() - t0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "599d9fcb-958b-4abb-ab1d-12761efbaaf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset.metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fdd5360-04ec-417c-a5f5-9f1fb8b1b6ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset.evaluate_segment_usage()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d69ca56b-8ec6-45df-98ec-a10b657f4469",
   "metadata": {},
   "outputs": [],
   "source": [
    "folder_to_run_mapping = {}\n",
    "for folder in dataset.metadata[['run', 'folder']].folder.unique():\n",
    "    folder_to_run_mapping[folder] = set(dataset.metadata[dataset.metadata.folder == folder].run)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ac8628f-5ec0-45c4-be96-cd99d339a808",
   "metadata": {},
   "outputs": [],
   "source": [
    "norm_usage_table_v = NormalizedSegmentUsageTable.load_from_repertoire_dataset(repertoire_dataset=dataset, gene='TRA', segment_type='V')\n",
    "norm_usage_table_j = NormalizedSegmentUsageTable.load_from_repertoire_dataset(repertoire_dataset=dataset, gene='TRA', segment_type='J')\n",
    "\n",
    "z_score_usage_table_v = StandardizedSegmentUsageTable.load_from_repertoire_dataset(repertoire_dataset=dataset, gene='TRA', segment_type='V', group_mapping=folder_to_run_mapping, standardization_method='log_exp')\n",
    "z_score_usage_table_j = StandardizedSegmentUsageTable.load_from_repertoire_dataset(repertoire_dataset=dataset, gene='TRA', segment_type='J', group_mapping=folder_to_run_mapping, standardization_method='log_exp')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b35b971-5746-4598-8c1d-c443cf5b62f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset[0].number_of_reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a32a755-bdb3-463d-9c4b-d3681c13751c",
   "metadata": {},
   "outputs": [],
   "source": [
    "norm_usage_table_v"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04b892cc-22e3-450f-a541-5a18cdc1c0fb",
   "metadata": {},
   "source": [
    "#### Sampling reads from the repertoire, not changing the gene usage distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4b599e3-5f45-4e4b-8df7-4a3ddf0163d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "sampled_repertoire = RepertoireSampling().sample(repertoire=dataset[0], old_usage_matrix=[norm_usage_table_v, norm_usage_table_j], new_usage_matrix=None, n=100000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a803686b-dd61-4676-b5bd-14dc9cd1725b",
   "metadata": {},
   "outputs": [],
   "source": [
    "sampled_repertoire"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "392e2fa3-12c2-4947-a7e1-33bb5d25b55e",
   "metadata": {},
   "source": [
    "#### Resampling the repertoire using the new derived usages for V and J genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7048be5f-1464-49da-b079-6cca46350adc",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1461e1b2-4c36-4d1d-86db-fc75a5faf2bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "t0 = time.time()\n",
    "resampled_repertoire = RepertoireSampling().sample(repertoire=dataset[0], old_usage_matrix=[norm_usage_table_v, norm_usage_table_j], new_usage_matrix=[z_score_usage_table_v, z_score_usage_table_j])\n",
    "print(time.time() - t0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "971d714e-8533-4b62-8990-e04d190196c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "resampled_repertoire"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "037fc09e-1076-4278-8939-ef60088f74d0",
   "metadata": {},
   "source": [
    "#### Resampling all the reperoires from a dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f294abf-cccb-46f4-8612-6dfef21f6cb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "t0 = time.time()\n",
    "resampled_dataset = dataset.resample(updated_segment_usage_tables=[z_score_usage_table_v, z_score_usage_table_j], threads=28)\n",
    "print(time.time() - t0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2dfa2c3d-6038-479a-b01c-5e48b489b0a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset[14]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82bdf1f0-7910-4418-9cf2-9e85a4e8f368",
   "metadata": {},
   "outputs": [],
   "source": [
    "resampled_dataset[14]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38abd9eb-2334-4097-8081-ea0d1dadbff8",
   "metadata": {},
   "outputs": [],
   "source": [
    "of_trav_matrix = NormalizedSegmentUsageTable.load_from_repertoire_dataset(\n",
    "    resampled_dataset, gene='TRA', segment_type='V').segment_usage_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dea46e11-dc03-4ab0-bea5-0d1761b2870d",
   "metadata": {},
   "outputs": [],
   "source": [
    "of_traj_matrix = NormalizedSegmentUsageTable.load_from_repertoire_dataset(\n",
    "    resampled_dataset, gene='TRA', segment_type='J').segment_usage_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce41713c-e61e-4e2f-be90-57cf621a8ae2",
   "metadata": {},
   "outputs": [],
   "source": [
    "of_traj_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9fea8c0-429e-4ddd-bbf4-42da971722e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "of_trav_matrix.to_csv('../../../data/cohort_I/onlyfunctional_corrected/of_v_clonotypes_TRA.csv', index = False)\n",
    "of_traj_matrix.to_csv('../../../data/cohort_I/onlyfunctional_corrected/of_j_clonotypes_TRA.csv', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b33cfe0d-8f92-4381-859a-df44f9414b25",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ec3d5d8-c9d4-4218-a89f-ceb740f8a93c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "028eb684-1d71-416c-9096-7612e0242af0",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Beta chain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ce9a84c-95e4-480d-a1fe-6523552adb1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#metadata\n",
    "metadata_path = '../../../data/cohort_I/metadata/metadata_fmba_full.txt'\n",
    "batch_meta_path = '../../../data/cohort_I/metadata/desc_fmba_not_nan_hla.csv'\n",
    "\n",
    "metadata = pd.read_csv(metadata_path, sep='\\t')\n",
    "metadata.rename(columns={'id': 'file.id_short'}, inplace=True)\n",
    "metadata['file.id_short'] = metadata['file.id_short'].astype(int)\n",
    "\n",
    "batch_meta = pd.read_csv(batch_meta_path, sep=',')\n",
    "batch_meta['file.id_short'] = batch_meta['file.id'].str.split('_').str[0]\n",
    "batch_meta['file.id_short'] = batch_meta['file.id_short'].astype(int)\n",
    "batch_meta['run'] = batch_meta['file.id'] + '.clonotypes.TRB.txt'\n",
    "\n",
    "new_meta = pd.merge(metadata, batch_meta[['file.id_short', 'run', 'folder']], on='file.id_short')\n",
    "new_meta = new_meta[new_meta['sample.COVID_status'] == 'healthy']\n",
    "new_meta = new_meta.reset_index(drop=True)\n",
    "new_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "602e87ed-0ac4-40d8-98e5-1502b41ab9ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "base_path = '../../../data/cohort_I/data/'\n",
    "t0 = time.time()\n",
    "dataset = RepertoireDataset.load(parser=VDJtoolsParser(sep=','), \n",
    "                                 metadata=new_meta,\n",
    "                                 threads=28,\n",
    "                                 gene = 'TRB',\n",
    "                                 paths=[base_path + r['run']\n",
    "                                        for _, r in new_meta.iterrows()])\n",
    "print(time.time() - t0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f5bbfa1-0c57-435b-8b73-3f543b910b1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset.metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ef2c1bc-48ad-4fb0-85c1-3b3108f11114",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset.evaluate_segment_usage()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eeacf81d-b680-4fa0-bcec-6fc21cd35b76",
   "metadata": {},
   "outputs": [],
   "source": [
    "folder_to_run_mapping = {}\n",
    "for folder in dataset.metadata[['run', 'folder']].folder.unique():\n",
    "    folder_to_run_mapping[folder] = set(dataset.metadata[dataset.metadata.folder == folder].run)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f38e496b-8166-4308-831a-582f6cffd8a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "norm_usage_table_v = NormalizedSegmentUsageTable.load_from_repertoire_dataset(repertoire_dataset=dataset, gene='TRB', segment_type='V')\n",
    "norm_usage_table_j = NormalizedSegmentUsageTable.load_from_repertoire_dataset(repertoire_dataset=dataset, gene='TRB', segment_type='J')\n",
    "\n",
    "z_score_usage_table_v = StandardizedSegmentUsageTable.load_from_repertoire_dataset(repertoire_dataset=dataset, gene='TRB', segment_type='V', group_mapping=folder_to_run_mapping, standardization_method='log_exp')\n",
    "z_score_usage_table_j = StandardizedSegmentUsageTable.load_from_repertoire_dataset(repertoire_dataset=dataset, gene='TRB', segment_type='J', group_mapping=folder_to_run_mapping, standardization_method='log_exp')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fe30c9f-8b87-44bc-9a02-bb7c074870e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "norm_usage_table_v"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb1e4c59-8651-460b-bdb2-0ee85eeacc24",
   "metadata": {},
   "source": [
    "#### Sampling reads from the repertoire, not changing the gene usage distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10e313d6-72d0-4049-ad54-608e416a8d11",
   "metadata": {},
   "outputs": [],
   "source": [
    "sampled_repertoire = RepertoireSampling().sample(repertoire=dataset[0], old_usage_matrix=[norm_usage_table_v, norm_usage_table_j], new_usage_matrix=None, n=100000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86f6904e-b577-45cc-aec3-57e47640640f",
   "metadata": {},
   "outputs": [],
   "source": [
    "sampled_repertoire"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f3b5369-6145-423c-a272-0a53d17b5230",
   "metadata": {},
   "source": [
    "#### Resampling the repertoire using the new derived usages for V and J genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf0db67e-91ba-4911-b6ea-a3411096b472",
   "metadata": {},
   "outputs": [],
   "source": [
    "t0 = time.time()\n",
    "resampled_repertoire = RepertoireSampling().sample(repertoire=dataset[0], old_usage_matrix=[norm_usage_table_v, norm_usage_table_j], new_usage_matrix=[z_score_usage_table_v, z_score_usage_table_j])\n",
    "print(time.time() - t0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cab9baf4-8e23-4c8a-b891-291bf4c58967",
   "metadata": {},
   "outputs": [],
   "source": [
    "resampled_repertoire"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac46fb9a-ea41-4506-b744-7c13cd0b9460",
   "metadata": {},
   "source": [
    "#### Resampling all the reperoires from a dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "712c4728-c488-4708-99a6-5b296fa6ebab",
   "metadata": {},
   "outputs": [],
   "source": [
    "t0 = time.time()\n",
    "resampled_dataset = dataset.resample(updated_segment_usage_tables=[z_score_usage_table_v, z_score_usage_table_j], threads=28)\n",
    "print(time.time() - t0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b20a8c4-fea2-49c2-833e-17c64f3b7cbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset[14]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9a4f033-8053-4afa-a2ff-b93c9933a24a",
   "metadata": {},
   "outputs": [],
   "source": [
    "resampled_dataset[14]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f864968a-1736-4a58-a097-57a106a57613",
   "metadata": {},
   "outputs": [],
   "source": [
    "trbv_matrix = NormalizedSegmentUsageTable.load_from_repertoire_dataset(\n",
    "    resampled_dataset, gene='TRB', segment_type='V').segment_usage_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5dcde7b6-1a2a-42a0-9c41-1b862e2a60c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "trBj_matrix = NormalizedSegmentUsageTable.load_from_repertoire_dataset(\n",
    "    resampled_dataset, gene='TRB', segment_type='J').segment_usage_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3647c886-df29-441e-80d8-e92835f883ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "trbj_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2ef85f4-45ae-4031-9a0b-8a13e220de82",
   "metadata": {},
   "outputs": [],
   "source": [
    "trbv_matrix.to_csv('../../../data/cohort_I/data_corrected/v_clonotypes_TRB.csv', index = False)\n",
    "trbj_matrix.to_csv('../../../data/cohort_I/data_corrected/j_clonotypes_TRB.csv', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aee90c0a-70f0-4df9-8ddf-29e9c298b631",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "35e698f6-4f07-424d-ae58-0abf505d873a",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Nonfunctional sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63a9191d-1da2-4873-a1dd-91348f01e176",
   "metadata": {},
   "outputs": [],
   "source": [
    "base_path = '../../../data/cohort_I/nonfunctional/'\n",
    "t0 = time.time()\n",
    "dataset = RepertoireDataset.load(parser=VDJtoolsParser(sep=','), \n",
    "                                 metadata=new_meta,\n",
    "                                 threads=28,\n",
    "                                 gene = 'TRB',\n",
    "                                 paths=[base_path + r['run']\n",
    "                                        for _, r in new_meta.iterrows()])\n",
    "print(time.time() - t0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25fe10b0-e7cd-4dda-b051-23d500c0ec77",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset.metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fdcc52b2-a369-45b1-a434-0f9b694955de",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset.evaluate_segment_usage()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "418c0a36-b8d7-49f6-80cf-aa725a537840",
   "metadata": {},
   "outputs": [],
   "source": [
    "folder_to_run_mapping = {}\n",
    "for folder in dataset.metadata[['run', 'folder']].folder.unique():\n",
    "    folder_to_run_mapping[folder] = set(dataset.metadata[dataset.metadata.folder == folder].run)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78e54a74-e7cc-48fa-a46a-552f7b0cd1a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "norm_usage_table_v = NormalizedSegmentUsageTable.load_from_repertoire_dataset(repertoire_dataset=dataset, gene='TRB', segment_type='V')\n",
    "norm_usage_table_j = NormalizedSegmentUsageTable.load_from_repertoire_dataset(repertoire_dataset=dataset, gene='TRB', segment_type='J')\n",
    "\n",
    "z_score_usage_table_v = StandardizedSegmentUsageTable.load_from_repertoire_dataset(repertoire_dataset=dataset, gene='TRB', segment_type='V', group_mapping=folder_to_run_mapping, standardization_method='log_exp')\n",
    "z_score_usage_table_j = StandardizedSegmentUsageTable.load_from_repertoire_dataset(repertoire_dataset=dataset, gene='TRB', segment_type='J', group_mapping=folder_to_run_mapping, standardization_method='log_exp')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01e55a49-db99-4dd7-936f-f6a054556f1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "norm_usage_table_v"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58a38185-a6ce-4a3f-9bb9-0c250a4d4edf",
   "metadata": {},
   "source": [
    "#### Sampling reads from the repertoire, not changing the gene usage distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6b68c00-f6b0-4128-9ecc-4c6ece59b4cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "sampled_repertoire = RepertoireSampling().sample(repertoire=dataset[0], old_usage_matrix=[norm_usage_table_v, norm_usage_table_j], new_usage_matrix=None, n=100000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c476d4f6-09ce-47c6-becd-05e694580e04",
   "metadata": {},
   "outputs": [],
   "source": [
    "sampled_repertoire"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56b1d3df-0bc6-4f59-b5b2-30110a5171f9",
   "metadata": {},
   "source": [
    "#### Resampling the repertoire using the new derived usages for V and J genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3713814a-a182-4ce2-a291-36be1b769a8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c1564d9-dc30-4f80-b2bb-d5188060de3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "t0 = time.time()\n",
    "resampled_repertoire = RepertoireSampling().sample(repertoire=dataset[0], old_usage_matrix=[norm_usage_table_v, norm_usage_table_j], new_usage_matrix=[z_score_usage_table_v, z_score_usage_table_j])\n",
    "print(time.time() - t0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b794d2a4-0b71-4550-83b5-4713d2b4914d",
   "metadata": {},
   "outputs": [],
   "source": [
    "resampled_repertoire"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a418515-db6f-4f78-9c23-3c9f13d2ed98",
   "metadata": {},
   "source": [
    "#### Resampling all the reperoires from a dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d38598e-6c67-47e0-b4d0-9c20699eb729",
   "metadata": {},
   "outputs": [],
   "source": [
    "t0 = time.time()\n",
    "resampled_dataset = dataset.resample(updated_segment_usage_tables=[z_score_usage_table_v, z_score_usage_table_j], threads=28)\n",
    "print(time.time() - t0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d274803-badd-4001-8cd9-5dc58d018f2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset[14]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02814f22-e4c7-4b8f-a64a-e9e78120e7e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "resampled_dataset[14]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2c09718-84f2-46d9-a33b-9f457a5b2045",
   "metadata": {},
   "outputs": [],
   "source": [
    "nf_trbv_matrix = NormalizedSegmentUsageTable.load_from_repertoire_dataset(\n",
    "    resampled_dataset, gene='TRB', segment_type='V').segment_usage_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc72db8c-6e4f-429c-8dc0-4d7acc3a8b19",
   "metadata": {},
   "outputs": [],
   "source": [
    "nf_trbj_matrix = NormalizedSegmentUsageTable.load_from_repertoire_dataset(\n",
    "    resampled_dataset, gene='TRB', segment_type='J').segment_usage_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9964c83-a5fa-4428-baa1-c768868dce67",
   "metadata": {},
   "outputs": [],
   "source": [
    "nf_trbj_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e088a6fd-fc1f-4a2e-8c79-8346a7e3fba4",
   "metadata": {},
   "outputs": [],
   "source": [
    "nf_trbv_matrix.to_csv('../../../data/cohort_I/nonfunctional_corrected/nf_v_clonotypes_TRB.csv', index = False)\n",
    "nf_trbj_matrix.to_csv('../../../data/cohort_I/nonfunctional_corrected/nf_j_clonotypes_TRB.csv', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97d5b761-7ac1-48a3-ac83-9653b1cd03b0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "09a76491-6b8c-4b29-b5ea-6d988be78b16",
   "metadata": {},
   "source": [
    "### Onlyfunctional sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b96cb940-3b9e-4607-9797-b123a6a4cf2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "base_path = '../../../data/cohort_I/onlyfunctional/'\n",
    "t0 = time.time()\n",
    "dataset = RepertoireDataset.load(parser=VDJtoolsParser(sep=','), \n",
    "                                 metadata=new_meta,\n",
    "                                 threads=28,\n",
    "                                 gene = 'TRB',\n",
    "                                 paths=[base_path + r['run']\n",
    "                                        for _, r in new_meta.iterrows()])\n",
    "print(time.time() - t0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a3e96c5-2455-4363-bb06-7ce4dac2b9a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset.metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec3b9fd5-1e5e-4792-96ae-cf1e62dd11ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset.evaluate_segment_usage()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2de03726-579e-4642-a30e-e4ca356c2cb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "folder_to_run_mapping = {}\n",
    "for folder in dataset.metadata[['run', 'folder']].folder.unique():\n",
    "    folder_to_run_mapping[folder] = set(dataset.metadata[dataset.metadata.folder == folder].run)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd3e8665-6de6-4c41-aa65-1e8b31cca208",
   "metadata": {},
   "outputs": [],
   "source": [
    "norm_usage_table_v = NormalizedSegmentUsageTable.load_from_repertoire_dataset(repertoire_dataset=dataset, gene='TRB', segment_type='V')\n",
    "norm_usage_table_j = NormalizedSegmentUsageTable.load_from_repertoire_dataset(repertoire_dataset=dataset, gene='TRB', segment_type='J')\n",
    "\n",
    "z_score_usage_table_v = StandardizedSegmentUsageTable.load_from_repertoire_dataset(repertoire_dataset=dataset, gene='TRB', segment_type='V', group_mapping=folder_to_run_mapping, standardization_method='log_exp')\n",
    "z_score_usage_table_j = StandardizedSegmentUsageTable.load_from_repertoire_dataset(repertoire_dataset=dataset, gene='TRB', segment_type='J', group_mapping=folder_to_run_mapping, standardization_method='log_exp')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07796032-0245-45cb-a5b9-181c1cba76df",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset[0].number_of_reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a1c0b31-61e2-4968-82b5-a8f02bf043d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "norm_usage_table_v"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7c6efe0-1b55-47cd-8e71-49b746bf41cf",
   "metadata": {},
   "source": [
    "#### Sampling reads from the repertoire, not changing the gene usage distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f98f2bee-587f-4841-8b5b-01854d30bb9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "sampled_repertoire = RepertoireSampling().sample(repertoire=dataset[0], old_usage_matrix=[norm_usage_table_v, norm_usage_table_j], new_usage_matrix=None, n=100000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dcddf8f9-9849-474d-a6a9-c18684b07ee0",
   "metadata": {},
   "outputs": [],
   "source": [
    "sampled_repertoire"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "700b46bf-311b-43cb-ac02-b7241c7f0e74",
   "metadata": {},
   "source": [
    "#### Resampling the repertoire using the new derived usages for V and J genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afc164e3-718d-4f50-9ac7-a09c8feb2fdc",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e721f83b-fae8-46ff-b821-fb5607740047",
   "metadata": {},
   "outputs": [],
   "source": [
    "t0 = time.time()\n",
    "resampled_repertoire = RepertoireSampling().sample(repertoire=dataset[0], old_usage_matrix=[norm_usage_table_v, norm_usage_table_j], new_usage_matrix=[z_score_usage_table_v, z_score_usage_table_j])\n",
    "print(time.time() - t0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0abbdd79-0d31-433c-b2fe-78bbc08b5740",
   "metadata": {},
   "outputs": [],
   "source": [
    "resampled_repertoire"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07087ff8-fd8b-476c-9a07-cba5eb89819b",
   "metadata": {},
   "source": [
    "#### Resampling all the reperoires from a dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d64631d-6d7e-4a7f-aa11-5bc67d578968",
   "metadata": {},
   "outputs": [],
   "source": [
    "t0 = time.time()\n",
    "resampled_dataset = dataset.resample(updated_segment_usage_tables=[z_score_usage_table_v, z_score_usage_table_j], threads=28)\n",
    "print(time.time() - t0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb16d48c-2619-4583-bf58-0c4a9261d4e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset[14]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c16d7fa-06c8-4101-be76-f83663b45377",
   "metadata": {},
   "outputs": [],
   "source": [
    "resampled_dataset[14]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e8fdde2-0edd-455f-9385-9e3d88fee8fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "of_trbv_matrix = NormalizedSegmentUsageTable.load_from_repertoire_dataset(\n",
    "    resampled_dataset, gene='TRB', segment_type='V').segment_usage_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f190acb-d510-4635-9bf3-88a600335129",
   "metadata": {},
   "outputs": [],
   "source": [
    "of_trbj_matrix = NormalizedSegmentUsageTable.load_from_repertoire_dataset(\n",
    "    resampled_dataset, gene='TRB', segment_type='J').segment_usage_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e72670c2-dbe0-4125-ae73-18579bb1c572",
   "metadata": {},
   "outputs": [],
   "source": [
    "of_trbj_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df4a18e3-f788-4d80-ba28-6b5fe991de6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "of_trbv_matrix.to_csv('../../../data/cohort_I/onlyfunctional_corrected/of_v_clonotypes_TRB.csv', index = False)\n",
    "of_trbj_matrix.to_csv('../../../data/cohort_I/onlyfunctional_corrected/of_j_clonotypes_TRB.csv', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f059315-f0c4-4807-85c1-89484e3271fe",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9b790c5-0ed3-49f0-b603-4fd5d30d28a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "base_path = '../../../data/cohort_I/nonfunctional/'#'/home/obagrova/TRB_onlyfunc_nonfunc/nonfunctional/data/'\n",
    "t0 = time.time()\n",
    "dataset = RepertoireDataset.load(parser=VDJtoolsParser(sep=','), \n",
    "                                 metadata=new_meta,\n",
    "                                 threads=28,\n",
    "                                 gene = 'TRB',\n",
    "                                 paths=[base_path + r['run']\n",
    "                                        for _, r in new_meta.iterrows()])\n",
    "print(time.time() - t0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33e49b00-c3b0-4880-bfcd-720429bd2a75",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset.metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8cded89-f520-42d4-b3fb-04ebd372879d",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset.evaluate_segment_usage()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b040e757-b95c-4bb7-a51c-6dc39a222ee2",
   "metadata": {},
   "outputs": [],
   "source": [
    "folder_to_run_mapping = {}\n",
    "for folder in dataset.metadata[['run', 'folder']].folder.unique():\n",
    "    folder_to_run_mapping[folder] = set(dataset.metadata[dataset.metadata.folder == folder].run)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75a59953-8f6c-4226-8db1-c72990485a54",
   "metadata": {},
   "outputs": [],
   "source": [
    "norm_usage_table_v = NormalizedSegmentUsageTable.load_from_repertoire_dataset(repertoire_dataset=dataset, gene='TRB', segment_type='V')\n",
    "norm_usage_table_j = NormalizedSegmentUsageTable.load_from_repertoire_dataset(repertoire_dataset=dataset, gene='TRB', segment_type='J')\n",
    "\n",
    "z_score_usage_table_v = StandardizedSegmentUsageTable.load_from_repertoire_dataset(repertoire_dataset=dataset, gene='TRB', segment_type='V', group_mapping=folder_to_run_mapping, standardization_method='log_exp')\n",
    "z_score_usage_table_j = StandardizedSegmentUsageTable.load_from_repertoire_dataset(repertoire_dataset=dataset, gene='TRB', segment_type='J', group_mapping=folder_to_run_mapping, standardization_method='log_exp')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b53e462-ab93-48eb-8c7d-7dfb5f180bae",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset[0].number_of_reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d37c2e69-9b9d-4337-98f1-6c1ba17c3166",
   "metadata": {},
   "outputs": [],
   "source": [
    "norm_usage_table_v"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53925c91-4239-4dfa-ad83-ea25ba101e35",
   "metadata": {},
   "source": [
    "#### Sampling reads from the repertoire, not changing the gene usage distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68fc857f-cc11-44f7-9722-fba695ce70f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "sampled_repertoire = RepertoireSampling().sample(repertoire=dataset[0], old_usage_matrix=[norm_usage_table_v, norm_usage_table_j], new_usage_matrix=None, n=100000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e4b8a9f-7ea1-42a5-8a16-63002e3f88e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "sampled_repertoire"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c21a033-80ba-4b19-80db-a0508b20d68b",
   "metadata": {},
   "source": [
    "#### Resampling the repertoire using the new derived usages for V and J genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc7107ec-2228-417e-a39e-eb6888c4aeaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f39ca2c-4c89-4370-9f83-1930fc89402b",
   "metadata": {},
   "outputs": [],
   "source": [
    "t0 = time.time()\n",
    "resampled_repertoire = RepertoireSampling().sample(repertoire=dataset[0], old_usage_matrix=[norm_usage_table_v, norm_usage_table_j], new_usage_matrix=[z_score_usage_table_v, z_score_usage_table_j])\n",
    "print(time.time() - t0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee552d06-3cff-4ed5-b195-99b743cd3aba",
   "metadata": {},
   "outputs": [],
   "source": [
    "resampled_repertoire"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3b06bf5-f31f-4207-ae69-d6cf31614188",
   "metadata": {},
   "source": [
    "#### Resampling all the reperoires from a dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f3ac64f-cfd4-4d4d-8ca0-67f9a2653c9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "t0 = time.time()\n",
    "resampled_dataset = dataset.resample(updated_segment_usage_tables=[z_score_usage_table_v, z_score_usage_table_j], threads=28)\n",
    "print(time.time() - t0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0df74b84-ffef-4c50-a3ef-15f7c2547e8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset[14]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff7b280f-bef6-4a27-ba3d-2ccf5217b5c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "resampled_dataset[14]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "256a2404-4d22-47cf-acd5-912743400d31",
   "metadata": {},
   "outputs": [],
   "source": [
    "nf_trbv_matrix = NormalizedSegmentUsageTable.load_from_repertoire_dataset(\n",
    "    resampled_dataset, gene='TRB', segment_type='V').segment_usage_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e63f4b6a-ea72-4f4a-ab7f-ad7c3877bd54",
   "metadata": {},
   "outputs": [],
   "source": [
    "nf_trbj_matrix = NormalizedSegmentUsageTable.load_from_repertoire_dataset(\n",
    "    resampled_dataset, gene='TRB', segment_type='J').segment_usage_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0eb4f8db-7d23-45ba-9d26-f4c2ecc6ea86",
   "metadata": {},
   "outputs": [],
   "source": [
    "nf_trbj_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "170e2309-4247-488f-a00d-18517a371337",
   "metadata": {},
   "outputs": [],
   "source": [
    "nf_trbv_matrix.to_csv('../../../data/cohort_I/nonfunctional_corrected/nf_v_clonotypes_TRB.csv', index = False)\n",
    "nf_trbj_matrix.to_csv('../../../data/cohort_I/nonfunctional_corrected/nf_j_clonotypes_TRB.csv', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b226a3fd-23f9-4c57-a605-2ec66b7c676f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "a753cfb0-f6a3-4db4-8477-de20b2c182fa",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-ob]",
   "language": "python",
   "name": "conda-env-.conda-ob-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
